{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2022.9.27'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from cdapython import Q, columns,unique_terms\n",
    "\n",
    "\n",
    "# cptacsubjectquery = Q('subject_associated_project = \"%cptac%\"')\n",
    "# df = cptacsubjectquery.subject.count.run(\n",
    "#     include=(\"specimen_id specimen_type specimen_Files\")\n",
    "# ).to_dataframe()\n",
    "\n",
    "\n",
    "# df\n",
    "Q.get_version()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Column_Name</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>GDC_FILTER</td>\n",
       "      <td>GDC filters applied universally across all MAFs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>GDC_Validation_Status</td>\n",
       "      <td>GDC implementation of validation checks. See n...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Matched_Norm_Sample_UUID</td>\n",
       "      <td>Unique GDC identifier for normal aliquot (1018...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>Tumor_Sample_UUID</td>\n",
       "      <td>Unique GDC identifier for tumor aliquot (10189...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>anatomical_site</td>\n",
       "      <td>Per GDC Dictionary, the text term that represe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>case_id</td>\n",
       "      <td>Unique GDC identifier for the underlying case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>fileUUID</td>\n",
       "      <td>|-delimited list of unique GDC identifiers for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>normal_bam_uuid</td>\n",
       "      <td>Unique GDC identifier for the underlying norma...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>species</td>\n",
       "      <td>The taxonomic group (e.g. species) of the pati...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>156</th>\n",
       "      <td>src_vcf_id</td>\n",
       "      <td>|-delimited list of GDC VCF file identifiers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>tumor_bam_uuid</td>\n",
       "      <td>Unique GDC identifier for the underlying bam file</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Column_Name  \\\n",
       "40                 GDC_FILTER   \n",
       "41      GDC_Validation_Status   \n",
       "54   Matched_Norm_Sample_UUID   \n",
       "78          Tumor_Sample_UUID   \n",
       "92            anatomical_site   \n",
       "97                    case_id   \n",
       "119                  fileUUID   \n",
       "133           normal_bam_uuid   \n",
       "149                   species   \n",
       "156                src_vcf_id   \n",
       "175            tumor_bam_uuid   \n",
       "\n",
       "                                           Description  \n",
       "40     GDC filters applied universally across all MAFs  \n",
       "41   GDC implementation of validation checks. See n...  \n",
       "54   Unique GDC identifier for normal aliquot (1018...  \n",
       "78   Unique GDC identifier for tumor aliquot (10189...  \n",
       "92   Per GDC Dictionary, the text term that represe...  \n",
       "97       Unique GDC identifier for the underlying case  \n",
       "119  |-delimited list of unique GDC identifiers for...  \n",
       "133  Unique GDC identifier for the underlying norma...  \n",
       "149  The taxonomic group (e.g. species) of the pati...  \n",
       "156       |-delimited list of GDC VCF file identifiers  \n",
       "175  Unique GDC identifier for the underlying bam file  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# col,val = \"Column_Name:specimen_Files\".split(\":\")\n",
    "# df = columns().to_dataframe()\n",
    "# df[df[col].str.contains(val, case=False, na=False)]\n",
    "\"\"\" to search with to_dataframe function you need the column name and the value use : to search like this Description:GDC \"\"\"\n",
    "columns().to_dataframe(include=\"Description:GDC\")        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GDC_FILTER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NonExonic;bitgt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NonExonic;bitgt;common_in_exac</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NonExonic;bitgt;common_in_exac;gdc_pon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NonExonic;bitgt;common_in_exac;gdc_pon;wga_pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NonExonic;bitgt;common_in_exac;wga_pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NonExonic;bitgt;gdc_pon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NonExonic;bitgt;gdc_pon;wga_normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NonExonic;bitgt;gdc_pon;wga_pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NonExonic;bitgt;ndp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NonExonic;bitgt;ndp;wga_pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NonExonic;bitgt;ndp;wga_pair|NonExonic;bitgt;w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NonExonic;bitgt;wga_normal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NonExonic;bitgt;wga_pair</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NonExonic;bitgt;wga_tumor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           GDC_FILTER\n",
       "1                                     NonExonic;bitgt\n",
       "2                      NonExonic;bitgt;common_in_exac\n",
       "3              NonExonic;bitgt;common_in_exac;gdc_pon\n",
       "4     NonExonic;bitgt;common_in_exac;gdc_pon;wga_pair\n",
       "5             NonExonic;bitgt;common_in_exac;wga_pair\n",
       "6                             NonExonic;bitgt;gdc_pon\n",
       "7                  NonExonic;bitgt;gdc_pon;wga_normal\n",
       "8                    NonExonic;bitgt;gdc_pon;wga_pair\n",
       "9                                 NonExonic;bitgt;ndp\n",
       "10                       NonExonic;bitgt;ndp;wga_pair\n",
       "11  NonExonic;bitgt;ndp;wga_pair|NonExonic;bitgt;w...\n",
       "12                         NonExonic;bitgt;wga_normal\n",
       "13                           NonExonic;bitgt;wga_pair\n",
       "14                          NonExonic;bitgt;wga_tumor"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"GDC_FILTER\").to_dataframe(include=\"GDC_FILTER:NonExonic;bitgt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "                        Total execution time: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">0</span>\n",
       "                        min <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3.688</span> sec <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3688</span> ms\n",
       "                        \n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n",
       "                        Total execution time: \u001b[1;36m0\u001b[0m\n",
       "                        min \u001b[1;36m3.688\u001b[0m sec \u001b[1;36m3688\u001b[0m ms\n",
       "                        \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_short_name</th>\n",
       "      <th>case_barcode</th>\n",
       "      <th>sample_barcode_tumor</th>\n",
       "      <th>sample_barcode_normal</th>\n",
       "      <th>aliquot_barcode_tumor</th>\n",
       "      <th>aliquot_barcode_normal</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>Center</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>...</th>\n",
       "      <th>tumor_bam_uuid</th>\n",
       "      <th>normal_bam_uuid</th>\n",
       "      <th>case_id</th>\n",
       "      <th>GDC_FILTER</th>\n",
       "      <th>COSMIC</th>\n",
       "      <th>MC3_Overlap</th>\n",
       "      <th>GDC_Validation_Status</th>\n",
       "      <th>callerName</th>\n",
       "      <th>fileUUID</th>\n",
       "      <th>fileName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-BRCA</td>\n",
       "      <td>TCGA-JL-A3YX</td>\n",
       "      <td>TCGA-JL-A3YX-01A</td>\n",
       "      <td>TCGA-JL-A3YX-10A</td>\n",
       "      <td>TCGA-JL-A3YX-01A-11D-A22X-09</td>\n",
       "      <td>TCGA-JL-A3YX-10A-01D-A22X-09</td>\n",
       "      <td>BLVRA</td>\n",
       "      <td>644</td>\n",
       "      <td>WUGSC</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>...</td>\n",
       "      <td>8e7c8c6a-e9ee-48cf-b919-e5bfa3e16f5c</td>\n",
       "      <td>b84f50ba-929d-4a86-8d52-ccdde2460fcb</td>\n",
       "      <td>75f6e476-233a-409c-a3ef-0240be569813</td>\n",
       "      <td>NonExonic;bitgt</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>muse|mutect|somaticsniper|varscan</td>\n",
       "      <td>b8ca5856-9819-459c-87c5-94e91aca4032|995c0111-...</td>\n",
       "      <td>TCGA.BRCA.muse.b8ca5856-9819-459c-87c5-94e91ac...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 109 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   project_short_name  case_barcode sample_barcode_tumor  \\\n",
       "96          TCGA-BRCA  TCGA-JL-A3YX     TCGA-JL-A3YX-01A   \n",
       "\n",
       "   sample_barcode_normal         aliquot_barcode_tumor  \\\n",
       "96      TCGA-JL-A3YX-10A  TCGA-JL-A3YX-01A-11D-A22X-09   \n",
       "\n",
       "          aliquot_barcode_normal Hugo_Symbol  Entrez_Gene_Id Center  \\\n",
       "96  TCGA-JL-A3YX-10A-01D-A22X-09       BLVRA             644  WUGSC   \n",
       "\n",
       "   NCBI_Build  ...                        tumor_bam_uuid  \\\n",
       "96     GRCh38  ...  8e7c8c6a-e9ee-48cf-b919-e5bfa3e16f5c   \n",
       "\n",
       "                         normal_bam_uuid  \\\n",
       "96  b84f50ba-929d-4a86-8d52-ccdde2460fcb   \n",
       "\n",
       "                                 case_id       GDC_FILTER COSMIC MC3_Overlap  \\\n",
       "96  75f6e476-233a-409c-a3ef-0240be569813  NonExonic;bitgt   None        True   \n",
       "\n",
       "   GDC_Validation_Status                         callerName  \\\n",
       "96               Unknown  muse|mutect|somaticsniper|varscan   \n",
       "\n",
       "                                             fileUUID  \\\n",
       "96  b8ca5856-9819-459c-87c5-94e91aca4032|995c0111-...   \n",
       "\n",
       "                                             fileName  \n",
       "96  TCGA.BRCA.muse.b8ca5856-9819-459c-87c5-94e91ac...  \n",
       "\n",
       "[1 rows x 109 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q(\"GDC_FILTER = 'NonExonic;bitgt'\").mutation.run().to_dataframe(include=\"case_barcode:TCGA-JL-\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.14 ('ven3': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.14"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "eae9c1a3dd8e2f898c643d98dd719bbe12e700aeabe7bf687912cfc443f15d3e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
