{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "614a478f",
   "metadata": {},
   "source": [
    "# Cohort Building"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8004e87",
   "metadata": {},
   "source": [
    "**Example use case:** \n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" alt=\"alt_text\" align=\"left\"\n",
    "\twidth=\"150\" height=\"150\" />\n",
    "Julia is an oncologist that specializes in female reproductive health. As part of her research, she is interested in  using existing data on uterine cancers. If possible, she would like to see multiple datatypes (gross imaging, genomic data, proteomic data, histology) that come from the same patient, so she can look for shared phenotypes to test for their potential as early diagnostics. Julia heard that the Cancer Data Aggregator has made it easy to search across multiple datasets created by NCI, and so has decided to start her search there.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e45a9b",
   "metadata": {},
   "source": [
    "## Getting Started\n",
    "\n",
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "\n",
    "Before Julia does any work, she needs to import these functions cdapython.\n",
    "She'll also need to import [pandas](https://pandas.pydata.org/) to get nice dataframes.\n",
    "Finally, she tells cdapython to report it's version so she can be sure she's using the one she means to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5265d4d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022.6.28\n"
     ]
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "import pandas as pd \n",
    "print(cdapython.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c577d416",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://cancerdata.dsde-dev.broadinstitute.org/'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q.set_default_project_dataset(\"broad-dsde-dev.cda_dev\")\n",
    "Q.set_host_url(\"https://cancerdata.dsde-dev.broadinstitute.org/\")\n",
    "Q.get_host_url()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7a4e1b8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'broad-dsde-dev.cda_dev'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "Q.get_default_project_dataset()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75eef23e",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>\n",
    "<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>\n",
    "<li><b>treatment:</b> Represent medication administration or other treatment types.</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "391bc9a7",
   "metadata": {},
   "source": [
    "## Finding Search Terms\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "   Accordingly, to see what search fields are available, Julia starts by using the command `columns`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ef0dd8e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['File.id',\n",
       " 'File.identifier.system',\n",
       " 'File.identifier.value',\n",
       " 'File.label',\n",
       " 'File.data_category',\n",
       " 'File.data_type',\n",
       " 'File.file_format',\n",
       " 'File.associated_project',\n",
       " 'File.drs_uri',\n",
       " 'File.byte_size',\n",
       " 'File.checksum',\n",
       " 'File.data_modality',\n",
       " 'File.imaging_modality',\n",
       " 'File.dbgap_accession_number',\n",
       " 'File.imaging_series',\n",
       " 'id',\n",
       " 'identifier.system',\n",
       " 'identifier.value',\n",
       " 'species',\n",
       " 'sex',\n",
       " 'race',\n",
       " 'ethnicity',\n",
       " 'days_to_birth',\n",
       " 'subject_associated_project',\n",
       " 'vital_status',\n",
       " 'days_to_death',\n",
       " 'cause_of_death',\n",
       " 'ResearchSubject.id',\n",
       " 'ResearchSubject.identifier.system',\n",
       " 'ResearchSubject.identifier.value',\n",
       " 'ResearchSubject.member_of_research_project',\n",
       " 'ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles',\n",
       " 'ResearchSubject.Specimen.id',\n",
       " 'ResearchSubject.Specimen.identifier.system',\n",
       " 'ResearchSubject.Specimen.identifier.value',\n",
       " 'ResearchSubject.Specimen.associated_project',\n",
       " 'ResearchSubject.Specimen.days_to_collection',\n",
       " 'ResearchSubject.Specimen.primary_disease_type',\n",
       " 'ResearchSubject.Specimen.anatomical_site',\n",
       " 'ResearchSubject.Specimen.source_material_type',\n",
       " 'ResearchSubject.Specimen.specimen_type',\n",
       " 'ResearchSubject.Specimen.derived_from_specimen',\n",
       " 'ResearchSubject.Specimen.derived_from_subject']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd05eba2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "There are a lot of columns in the CDA data, but Julia is most interested in diagnosis data, so she filters the list to only those:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "536970c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ResearchSubject.primary_diagnosis_condition',\n",
       " 'ResearchSubject.primary_diagnosis_site',\n",
       " 'ResearchSubject.Diagnosis.id',\n",
       " 'ResearchSubject.Diagnosis.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.primary_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.age_at_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.morphology',\n",
       " 'ResearchSubject.Diagnosis.stage',\n",
       " 'ResearchSubject.Diagnosis.grade',\n",
       " 'ResearchSubject.Diagnosis.method_of_diagnosis',\n",
       " 'ResearchSubject.Diagnosis.Treatment.id',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.system',\n",
       " 'ResearchSubject.Diagnosis.Treatment.identifier.value',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_type',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_outcome',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_start',\n",
       " 'ResearchSubject.Diagnosis.Treatment.days_to_treatment_end',\n",
       " 'ResearchSubject.Diagnosis.Treatment.therapeutic_agent',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_effect',\n",
       " 'ResearchSubject.Diagnosis.Treatment.treatment_end_reason',\n",
       " 'ResearchSubject.Diagnosis.Treatment.number_of_cycles']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list(filters=\"diagnosis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63b4cf0",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "\n",
    "To search the CDA, a user also needs to know what search terms are available. Each column will contain a huge amount of data, so retreiving all of the rows would be overwhelming. Instead, the CDA has a `unique_terms()` function that will return all of the unique values that populate the requested column. Like `columns`, `unique_terms` defaults to giving us an overview of the results, and can be filtered.\n",
    "    \n",
    "</div>\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Since Julia is interested specificially in uterine cancers, she uses the `unique_terms` function to see what data is available for 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site' and 'ResearchSubject.primary_diagnosis_site' to see if 'uterine' appears:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4527dde5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Brain',\n",
       " 'Cervix',\n",
       " 'Head - Face Or Neck, Nos',\n",
       " 'Lymph Node(s) Paraaortic',\n",
       " 'Other',\n",
       " 'Pelvis',\n",
       " 'Spine',\n",
       " 'Unknown']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site\").to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "740e5955",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Abdomen',\n",
       " 'Abdomen, Mediastinum',\n",
       " 'Abdomen, Pelvis',\n",
       " 'Adrenal Glands',\n",
       " 'Adrenal gland',\n",
       " 'Anus and anal canal',\n",
       " 'Base of tongue',\n",
       " 'Bile Duct',\n",
       " 'Bladder',\n",
       " 'Bones, joints and articular cartilage of limbs',\n",
       " 'Bones, joints and articular cartilage of other and unspecified sites',\n",
       " 'Brain',\n",
       " 'Breast',\n",
       " 'Bronchus and lung',\n",
       " 'Cervix',\n",
       " 'Cervix uteri',\n",
       " 'Chest',\n",
       " 'Chest-Abdomen-Pelvis, Leg, TSpine',\n",
       " 'Colon',\n",
       " 'Connective, subcutaneous and other soft tissues',\n",
       " 'Corpus uteri',\n",
       " 'Ear',\n",
       " 'Esophagus',\n",
       " 'Extremities',\n",
       " 'Eye and adnexa',\n",
       " 'Floor of mouth',\n",
       " 'Gallbladder',\n",
       " 'Gum',\n",
       " 'Head',\n",
       " 'Head and Neck',\n",
       " 'Head-Neck',\n",
       " 'Heart, mediastinum, and pleura',\n",
       " 'Hematopoietic and reticuloendothelial systems',\n",
       " 'Hypopharynx',\n",
       " 'Intraocular',\n",
       " 'Kidney',\n",
       " 'Larynx',\n",
       " 'Lip',\n",
       " 'Liver',\n",
       " 'Liver and intrahepatic bile ducts',\n",
       " 'Lung',\n",
       " 'Lung Phantom',\n",
       " 'Lymph nodes',\n",
       " 'Marrow, Blood',\n",
       " 'Meninges',\n",
       " 'Mesothelium',\n",
       " 'Nasal cavity and middle ear',\n",
       " 'Nasopharynx',\n",
       " 'Not Reported',\n",
       " 'Oropharynx',\n",
       " 'Other and ill-defined digestive organs',\n",
       " 'Other and ill-defined sites',\n",
       " 'Other and ill-defined sites in lip, oral cavity and pharynx',\n",
       " 'Other and ill-defined sites within respiratory system and intrathoracic organs',\n",
       " 'Other and unspecified female genital organs',\n",
       " 'Other and unspecified major salivary glands',\n",
       " 'Other and unspecified male genital organs',\n",
       " 'Other and unspecified parts of biliary tract',\n",
       " 'Other and unspecified parts of mouth',\n",
       " 'Other and unspecified parts of tongue',\n",
       " 'Other and unspecified urinary organs',\n",
       " 'Other endocrine glands and related structures',\n",
       " 'Ovary',\n",
       " 'Palate',\n",
       " 'Pancreas',\n",
       " 'Pancreas ',\n",
       " 'Pelvis, Prostate, Anus',\n",
       " 'Penis',\n",
       " 'Peripheral nerves and autonomic nervous system',\n",
       " 'Phantom',\n",
       " 'Prostate',\n",
       " 'Prostate gland',\n",
       " 'Rectosigmoid junction',\n",
       " 'Rectum',\n",
       " 'Renal pelvis',\n",
       " 'Retroperitoneum and peritoneum',\n",
       " 'Skin',\n",
       " 'Small intestine',\n",
       " 'Spinal cord, cranial nerves, and other parts of central nervous system',\n",
       " 'Stomach',\n",
       " 'Testicles',\n",
       " 'Testis',\n",
       " 'Thymus',\n",
       " 'Thyroid',\n",
       " 'Thyroid gland',\n",
       " 'Tonsil',\n",
       " 'Trachea',\n",
       " 'Unknown',\n",
       " 'Ureter',\n",
       " 'Uterus',\n",
       " 'Uterus, NOS',\n",
       " 'Vagina',\n",
       " 'Various',\n",
       " 'Various (11 locations)',\n",
       " 'Vulva']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b005036b",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "CDA makes multiple datasets searchable from a common interface, but does not harmonize the data. This means that researchers should review all the terms in a column, and not just choose the first one that fits, as there may be other similar terms available as well.\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e6b8dc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia sees that \"treatment_anatomic_site\" does not have 'Uterine', but does have 'Cervix'. She also notes that both 'Uterus' and 'Uterus, NOS' are listed in the \"primary_diagnosis_site\" results. As she was initially looking for \"uterine\", Julia decides to expand her search a bit to account for variable naming schemes. So, she runs a fuzzy match filter on the \"ResearchSubject.primary_diagnosis_site\" for 'uter' as that should cover all variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "31064125",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix uteri', 'Corpus uteri', 'Uterus', 'Uterus, NOS']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"uter\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9311a49e",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Just to be sure, Julia also searches for any other instances of \"cervix\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2038a8cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Cervix', 'Cervix uteri']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"cerv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29c4de58",
   "metadata": {},
   "source": [
    "## Building a Query\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "With all her likely terms found, Julia begins to create a search that will get data for all of her terms. She does this by writing a series of `Q` statements that define what rows should be returned from each column. For the \"treatment_anatomic_site\", only one term is of interest, so she uses the `=` operator to get only exact matches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "951fcc8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Tsite = Q('ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site = \"Cervix\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cb5f72",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "However, for \"primary_diagnosis_site\", Julia has several terms she wants to search with. Luckily, `Q` also can run fuzzy searches. It can also search more than one term at a time, so Julia writes one big `Q` statement to grab everything that is either 'uter' or 'cerv':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "36cfd8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Dsite = Q('ResearchSubject.primary_diagnosis_site = \"%uter%\" OR ResearchSubject.primary_diagnosis_site = \"%cerv%\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "349af6f2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia adds her two queries together into one large one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9f5f9e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDATA = Tsite.OR(Dsite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1f5cb55",
   "metadata": {},
   "source": [
    "## Looking at Summary Data\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Now that Julia has a query, she can use it to look for data in any of the CDA endpoints. She starts by getting an overall summary of what data is available using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "355b1706",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4849 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">specimen_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">40793</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "specimen_count : \u001b[1;36m40793\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">treatment_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3049</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "treatment_count : \u001b[1;36m3049\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">diagnosis_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3685</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "diagnosis_count : \u001b[1;36m3685\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">researchsubject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4869</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "researchsubject_count : \u001b[1;36m4869\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">subject_count : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">3742</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "subject_count : \u001b[1;36m3742\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": []
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ce25fc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "It seems there's a lot of data that might work for Julias study! Since she is interested in the beginings of cancer, she decides to start by looking at  the researchsubject information, since that is where most of the diagnosis information is. She again gets a summary using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "55b0cdeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 10077 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">    total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">4869</span>    \n",
       "</pre>\n"
      ],
      "text/plain": [
       "    total : \u001b[1;36m4869\u001b[0m    \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   files : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">324646</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   files : \u001b[1;36m324646\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_840c4_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_840c4_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_840c4_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_840c4_row0_col0\" class=\"data row0 col0\" >PDC</td>\n",
       "      <td id=\"T_840c4_row0_col1\" class=\"data row0 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_840c4_row1_col0\" class=\"data row1 col0\" >GDC</td>\n",
       "      <td id=\"T_840c4_row1_col1\" class=\"data row1 col1\" >3591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_840c4_row2_col0\" class=\"data row2 col0\" >IDC</td>\n",
       "      <td id=\"T_840c4_row2_col1\" class=\"data row2 col1\" >1174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_9d42c_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_9d42c_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_9d42c_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_condition</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row0_col0\" class=\"data row0 col0\" >Adenomas and Adenocarcinomas</td>\n",
       "      <td id=\"T_9d42c_row0_col1\" class=\"data row0 col1\" >1672</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row1_col0\" class=\"data row1 col0\" >Uterine Corpus Endometrial Carcinoma</td>\n",
       "      <td id=\"T_9d42c_row1_col1\" class=\"data row1 col1\" >104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row2_col0\" class=\"data row2 col0\" >Myomatous Neoplasms</td>\n",
       "      <td id=\"T_9d42c_row2_col1\" class=\"data row2 col1\" >188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row3_col0\" class=\"data row3 col0\" >Squamous Cell Neoplasms</td>\n",
       "      <td id=\"T_9d42c_row3_col1\" class=\"data row3 col1\" >609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row4_col0\" class=\"data row4 col0\" >Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td id=\"T_9d42c_row4_col1\" class=\"data row4 col1\" >487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row5_col0\" class=\"data row5 col0\" >Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td id=\"T_9d42c_row5_col1\" class=\"data row5 col1\" >320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row6_col0\" class=\"data row6 col0\" >Not Reported</td>\n",
       "      <td id=\"T_9d42c_row6_col1\" class=\"data row6 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row7_col0\" class=\"data row7 col0\" >None</td>\n",
       "      <td id=\"T_9d42c_row7_col1\" class=\"data row7 col1\" >1175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row8_col0\" class=\"data row8 col0\" >Complex Epithelial Neoplasms</td>\n",
       "      <td id=\"T_9d42c_row8_col1\" class=\"data row8 col1\" >27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row9_col0\" class=\"data row9 col0\" >Epithelial Neoplasms, NOS</td>\n",
       "      <td id=\"T_9d42c_row9_col1\" class=\"data row9 col1\" >230</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row10_col0\" class=\"data row10 col0\" >Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td id=\"T_9d42c_row10_col1\" class=\"data row10 col1\" >14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row11_col0\" class=\"data row11 col0\" >Trophoblastic neoplasms</td>\n",
       "      <td id=\"T_9d42c_row11_col1\" class=\"data row11 col1\" >13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row12_col0\" class=\"data row12 col0\" >Neoplasms, NOS</td>\n",
       "      <td id=\"T_9d42c_row12_col1\" class=\"data row12 col1\" >12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row13_col0\" class=\"data row13 col0\" >Mesonephromas</td>\n",
       "      <td id=\"T_9d42c_row13_col1\" class=\"data row13 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_9d42c_row14_col0\" class=\"data row14 col0\" >Neuroepitheliomatous Neoplasms</td>\n",
       "      <td id=\"T_9d42c_row14_col1\" class=\"data row14 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_02069_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_02069_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_02069_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >primary_diagnosis_site</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_02069_row0_col0\" class=\"data row0 col0\" >Cervix uteri</td>\n",
       "      <td id=\"T_02069_row0_col1\" class=\"data row0 col1\" >915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_02069_row1_col0\" class=\"data row1 col0\" >Uterus, NOS</td>\n",
       "      <td id=\"T_02069_row1_col1\" class=\"data row1 col1\" >2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_02069_row2_col0\" class=\"data row2 col0\" >Corpus uteri</td>\n",
       "      <td id=\"T_02069_row2_col1\" class=\"data row2 col1\" >780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_02069_row3_col0\" class=\"data row3 col0\" >Uterus</td>\n",
       "      <td id=\"T_02069_row3_col1\" class=\"data row3 col1\" >867</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_02069_row4_col0\" class=\"data row4 col0\" >Cervix</td>\n",
       "      <td id=\"T_02069_row4_col1\" class=\"data row4 col1\" >307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ALLDATA.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a323e2",
   "metadata": {},
   "source": [
    "## Refining Queries\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Browsing the primary_diagnosis_condition data, Julia notices that there are a large number of research subjects that are Adenomas and Adenocarcinomas. Since Julia wants to look for common phenotypes in early cancers, she decides it might be easier to exclude the endocrine related data, as they might have different mechanisms. So she adds a new filter to her query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0d526198",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'ALLDATA' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_92/1888651251.py\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mNoadeno\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mQ\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'ResearchSubject.primary_diagnosis_condition != \"Adenomas and Adenocarcinomas\"'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mNoAdenoData\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mALLDATA\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAND\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mNoadeno\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mNoAdenoData\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresearchsubject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'ALLDATA' is not defined"
     ]
    }
   ],
   "source": [
    "Noadeno = Q('ResearchSubject.primary_diagnosis_condition != \"Adenomas and Adenocarcinomas\"')\n",
    "\n",
    "NoAdenoData = ALLDATA.AND(Noadeno)\n",
    "\n",
    "NoAdenoData.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a0191d",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "She then previews the actual metadata for researchsubject, subject, and file, to make sure that they have all the information she will need for her work. Since she's mostly interested in looking at the kinds of data available from each endpoint:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d186b837",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 4063 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0609d8ac-fc18-4e07-b5fb-12b372151dcf</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0609d8ac-fc18-4e0...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-011667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1d2937e8-6104-4330-b7dc-7bcd79dac927</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1d2937e8-6104-433...</td>\n",
       "      <td>TCGA-UCS</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>TCGA-N5-A4RN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2639fc46-2bf7-4050-b5dd-c4d8beb89d2e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2639fc46-2bf7-405...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0003476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30a4fa5d-98f4-4592-868f-2beb443eb82d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '30a4fa5d-98f4-459...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD13447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30ffb337-c844-493f-97d6-7aea157d1b5e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '30ffb337-c844-493...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>TCGA-N8-A4PN__tcga_ucs</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-N8-A4PN__tcg...</td>\n",
       "      <td>tcga_ucs</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>TCGA-N8-A4PN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>TCGA-ND-A4WA__tcga_ucs</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-ND-A4WA__tcg...</td>\n",
       "      <td>tcga_ucs</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>TCGA-ND-A4WA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>a6a07bc7-ab1b-429d-81ed-d7a890c88e51</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a6a07bc7-ab1b-429...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-VS-A9V1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>a7f3f44f-5d75-4572-80e6-279f186e44a3</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'a7f3f44f-5d75-457...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Epithelial Neoplasms, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0018781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>cca23650-1b06-4977-91e8-f35e73e70c48</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'cca23650-1b06-497...</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>TCGA-VS-A9UC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   0609d8ac-fc18-4e07-b5fb-12b372151dcf   \n",
       "1   1d2937e8-6104-4330-b7dc-7bcd79dac927   \n",
       "2   2639fc46-2bf7-4050-b5dd-c4d8beb89d2e   \n",
       "3   30a4fa5d-98f4-4592-868f-2beb443eb82d   \n",
       "4   30ffb337-c844-493f-97d6-7aea157d1b5e   \n",
       "..                                   ...   \n",
       "95                TCGA-N8-A4PN__tcga_ucs   \n",
       "96                TCGA-ND-A4WA__tcga_ucs   \n",
       "97  a6a07bc7-ab1b-429d-81ed-d7a890c88e51   \n",
       "98  a7f3f44f-5d75-4572-80e6-279f186e44a3   \n",
       "99  cca23650-1b06-4977-91e8-f35e73e70c48   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '0609d8ac-fc18-4e0...   \n",
       "1   [{'system': 'GDC', 'value': '1d2937e8-6104-433...   \n",
       "2   [{'system': 'GDC', 'value': '2639fc46-2bf7-405...   \n",
       "3   [{'system': 'GDC', 'value': '30a4fa5d-98f4-459...   \n",
       "4   [{'system': 'GDC', 'value': '30ffb337-c844-493...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'IDC', 'value': 'TCGA-N8-A4PN__tcg...   \n",
       "96  [{'system': 'IDC', 'value': 'TCGA-ND-A4WA__tcg...   \n",
       "97  [{'system': 'GDC', 'value': 'a6a07bc7-ab1b-429...   \n",
       "98  [{'system': 'GDC', 'value': 'a7f3f44f-5d75-457...   \n",
       "99  [{'system': 'GDC', 'value': 'cca23650-1b06-497...   \n",
       "\n",
       "   member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                  GENIE-DFCI   Soft Tissue Tumors and Sarcomas, NOS   \n",
       "1                    TCGA-UCS    Complex Mixed and Stromal Neoplasms   \n",
       "2                   GENIE-MSK                    Myomatous Neoplasms   \n",
       "3                       FM-AD  Cystic, Mucinous and Serous Neoplasms   \n",
       "4               CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "..                        ...                                    ...   \n",
       "95                   tcga_ucs                                   None   \n",
       "96                   tcga_ucs                                   None   \n",
       "97                  TCGA-CESC  Cystic, Mucinous and Serous Neoplasms   \n",
       "98                  GENIE-MSK              Epithelial Neoplasms, NOS   \n",
       "99                  TCGA-CESC                Squamous Cell Neoplasms   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0             Uterus, NOS    GENIE-DFCI-011667  \n",
       "1             Uterus, NOS         TCGA-N5-A4RN  \n",
       "2             Uterus, NOS  GENIE-MSK-P-0003476  \n",
       "3             Uterus, NOS              AD13447  \n",
       "4            Cervix uteri    HTMCP-03-06-02240  \n",
       "..                    ...                  ...  \n",
       "95                 Uterus         TCGA-N8-A4PN  \n",
       "96                 Uterus         TCGA-ND-A4WA  \n",
       "97           Cervix uteri         TCGA-VS-A9V1  \n",
       "98            Uterus, NOS  GENIE-MSK-P-0018781  \n",
       "99           Cervix uteri         TCGA-VS-A9UC  \n",
       "\n",
       "[100 rows x 6 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.researchsubject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "086697b3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For CDA, this is case_id.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>member_of_research_project:</b> A reference to the Study(s) of which this ResearchSubject is a member.</li>\n",
    "<li><b>primary_diagnosis_condition:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>primary_diagnosis_site:</b> The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This categorization groups cases into general categories. This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>subject_id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "8d0f5e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3838 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD15141</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD15141'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD290</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD290'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AD4252</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD4252'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AD9808</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD9808'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-02382</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-02382'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>GENIE-MSK-P-0001439</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00014...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-28854.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>GENIE-MSK-P-0001606</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00016...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-21549.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>GENIE-MSK-P-0001989</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00019...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-14610.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>GENIE-MSK-P-0013913</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00139...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-20454.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>GENIE-MSK-P-0015139</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00151...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-26298.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     id                                         identifier  \\\n",
       "0               AD15141            [{'system': 'GDC', 'value': 'AD15141'}]   \n",
       "1                 AD290              [{'system': 'GDC', 'value': 'AD290'}]   \n",
       "2                AD4252             [{'system': 'GDC', 'value': 'AD4252'}]   \n",
       "3                AD9808             [{'system': 'GDC', 'value': 'AD9808'}]   \n",
       "4             C3L-02382          [{'system': 'IDC', 'value': 'C3L-02382'}]   \n",
       "..                  ...                                                ...   \n",
       "95  GENIE-MSK-P-0001439  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00014...   \n",
       "96  GENIE-MSK-P-0001606  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00016...   \n",
       "97  GENIE-MSK-P-0001989  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00019...   \n",
       "98  GENIE-MSK-P-0013913  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00139...   \n",
       "99  GENIE-MSK-P-0015139  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00151...   \n",
       "\n",
       "         species     sex          race               ethnicity  days_to_birth  \\\n",
       "0   homo sapiens  female  not reported            not reported            NaN   \n",
       "1   homo sapiens  female  not reported            not reported            NaN   \n",
       "2   homo sapiens  female  not reported            not reported            NaN   \n",
       "3   homo sapiens  female  not reported            not reported            NaN   \n",
       "4   Homo sapiens    None          None                    None            NaN   \n",
       "..           ...     ...           ...                     ...            ...   \n",
       "95  homo sapiens  female         white  not hispanic or latino       -28854.0   \n",
       "96  homo sapiens  female         white  not hispanic or latino       -21549.0   \n",
       "97  homo sapiens  female         white  not hispanic or latino       -14610.0   \n",
       "98  homo sapiens  female         white  not hispanic or latino       -20454.0   \n",
       "99  homo sapiens  female         white  not hispanic or latino       -26298.0   \n",
       "\n",
       "   subject_associated_project  vital_status  days_to_death cause_of_death  \n",
       "0                     [FM-AD]  Not Reported            NaN           None  \n",
       "1                     [FM-AD]  Not Reported            NaN           None  \n",
       "2                     [FM-AD]  Not Reported            NaN           None  \n",
       "3                     [FM-AD]  Not Reported            NaN           None  \n",
       "4                [cptac_ucec]          None            NaN           None  \n",
       "..                        ...           ...            ...            ...  \n",
       "95                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "96                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "97                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "98                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "99                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.subject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec76132",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.\",STRING</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>species:</b> The taxonomic group (e.g. species) of the patient. For MVP, since taxonomy vocabulary is consistent between GDC and PDC, using text. Ultimately, this will be a term returned by the vocabulary service.</li>\n",
    "<li><b>sex:</b> The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics.</li>\n",
    "<li><b>race:</b> An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>ethnicity:</b> An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.</li>\n",
    "<li><b>subject_associated_project:</b> The list of Projects associated with the Subject.</li>\n",
    "<li><b>vital_status:</b> Coded value indicating the state or condition of being living or deceased; also includes the case where the vital status is unknown.</li>\n",
    "<li><b>days_to_death:</b> Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.</li>\n",
    "<li><b>cause_of_death:</b> Coded value indicating the circumstance or condition that results in the death of the subject.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "04e04136",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 21458 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>data_modality</th>\n",
       "      <th>imaging_modality</th>\n",
       "      <th>dbgap_accession_number</th>\n",
       "      <th>imaging_series</th>\n",
       "      <th>researchsubject_specimen_id</th>\n",
       "      <th>researchsubject_id</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>002ff7dd-d55c-4f60-8d9a-c814443ed665</td>\n",
       "      <td>[{'system': 'GDC', 'value': '002ff7dd-d55c-4f6...</td>\n",
       "      <td>9e4a57ec-ca07-4843-b77f-eb5504a817c4.rna_seq.c...</td>\n",
       "      <td>Sequencing Reads</td>\n",
       "      <td>Aligned Reads</td>\n",
       "      <td>BAM</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:002ff7dd-d55c-4f60-8d9a-c814443e...</td>\n",
       "      <td>96671613</td>\n",
       "      <td>6a4989df92ba58fe6558e7f6d0d5eb0f</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>a36de57d-2315-4bc3-93bf-69ec4f51063b</td>\n",
       "      <td>0bd52d75-5113-4ef1-bc75-509f59eaef2b</td>\n",
       "      <td>TCGA-JW-A5VG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>00305181-fec7-41c6-a143-f3bf4636bd5f</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00305181-fec7-41c...</td>\n",
       "      <td>39550f2d-268b-47a8-996c-7638f07de1dd.wxs.muse....</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Raw Simple Somatic Mutation</td>\n",
       "      <td>VCF</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:00305181-fec7-41c6-a143-f3bf4636...</td>\n",
       "      <td>25035</td>\n",
       "      <td>42b1a48e34690de6cd6851ca753cff9d</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>987af661-a8b7-47b0-800f-a0b6a3ef0096</td>\n",
       "      <td>e31db89c-15ac-45a4-9788-afb339b0f3bc</td>\n",
       "      <td>TCGA-AX-A3G6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>005ff5ca-0975-4b0b-b413-4a906701e923</td>\n",
       "      <td>[{'system': 'GDC', 'value': '005ff5ca-0975-4b0...</td>\n",
       "      <td>f059db9c-510f-4dc0-bf36-6038149bf06a.wxs.somat...</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Raw Simple Somatic Mutation</td>\n",
       "      <td>VCF</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:005ff5ca-0975-4b0b-b413-4a906701...</td>\n",
       "      <td>268168</td>\n",
       "      <td>4174aa64898dfbdf2a73634c73b77d01</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>8a53198f-ee29-4cca-8cb2-9944790bcc53</td>\n",
       "      <td>e15d92a1-6925-4b58-a494-d6f665f67b83</td>\n",
       "      <td>TCGA-LP-A4AU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>007c3c99-c50f-4e45-907a-7ab78f47ff0d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '007c3c99-c50f-4e4...</td>\n",
       "      <td>TCGA_UCS.7ea16309-04b4-41ff-93aa-dc3605be9fc7....</td>\n",
       "      <td>Simple Nucleotide Variation</td>\n",
       "      <td>Annotated Somatic Mutation</td>\n",
       "      <td>VCF</td>\n",
       "      <td>TCGA-UCS</td>\n",
       "      <td>drs://dg.4DFC:007c3c99-c50f-4e45-907a-7ab78f47...</td>\n",
       "      <td>229002</td>\n",
       "      <td>8d2de6c206adef0e1236a1008d9234f9</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>e8dc4a59-fdfb-40d0-afc9-f7f21eb0f339</td>\n",
       "      <td>a1206473-cf9d-4bca-97c9-67f35817b806</td>\n",
       "      <td>TCGA-N9-A4Q1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>00a7b64e-ffd2-11e8-abd7-005056921935</td>\n",
       "      <td>[{'system': 'PDC', 'value': '00a7b64e-ffd2-11e...</td>\n",
       "      <td>15CPTAC_UCEC_P_PNNL_20180503_B4S3_f08.mzid.gz</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzIdentML</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:00a7b64e-ffd2-11e8-abd7-00505692...</td>\n",
       "      <td>5832060</td>\n",
       "      <td>194e198498ab1c3ff18984bff2ef28a9</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>61c4d96e-1259-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>7ef278e9-118a-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>C3L-01925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>6a95c4ba-ffb4-11e8-95a3-005056921935</td>\n",
       "      <td>[{'system': 'PDC', 'value': '6a95c4ba-ffb4-11e...</td>\n",
       "      <td>07CPTAC_UCEC_W_PNNL_20170922_B2S3_f18.mzML.gz</td>\n",
       "      <td>Processed Mass Spectra</td>\n",
       "      <td>Open Standard</td>\n",
       "      <td>mzML</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:6a95c4ba-ffb4-11e8-95a3-00505692...</td>\n",
       "      <td>231084046</td>\n",
       "      <td>4eccdcd78267609a2e6e0cdd6d1f075f</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>c8a6ef50-1259-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>bf1f593d-118a-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>C3N-01217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>6afce870-f915-11e8-90c1-005056921935</td>\n",
       "      <td>[{'system': 'PDC', 'value': '6afce870-f915-11e...</td>\n",
       "      <td>10CPTAC_UCEC_P_PNNL_20180222_B3S2_f04.raw</td>\n",
       "      <td>Raw Mass Spectra</td>\n",
       "      <td>Proprietary</td>\n",
       "      <td>vendor-specific</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:6afce870-f915-11e8-90c1-00505692...</td>\n",
       "      <td>806180410</td>\n",
       "      <td>625413aa5fece36dc7fa5375f92580ec</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>d799fbea-1271-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>c289029e-118a-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>C3N-01267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>6b78fe54-ffcf-11e8-9f26-005056921935</td>\n",
       "      <td>[{'system': 'PDC', 'value': '6b78fe54-ffcf-11e...</td>\n",
       "      <td>03CPTAC_UCEC_P_PNNL_20170922_B1S3_f05.psm</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Text</td>\n",
       "      <td>tsv</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:6b78fe54-ffcf-11e8-9f26-00505692...</td>\n",
       "      <td>2957979</td>\n",
       "      <td>71d54336cc42e41228484d9399d22f59</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>39743eb5-1259-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>6b043589-118a-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>C3L-01256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>6bcbea24-ffcf-11e8-9f26-005056921935</td>\n",
       "      <td>[{'system': 'PDC', 'value': '6bcbea24-ffcf-11e...</td>\n",
       "      <td>03CPTAC_UCEC_P_PNNL_20170922_B1S3_f08.psm</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Text</td>\n",
       "      <td>tsv</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:6bcbea24-ffcf-11e8-9f26-00505692...</td>\n",
       "      <td>2496941</td>\n",
       "      <td>bb2eb73c4d58f988538ca5b6b8453cfb</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>bb56d00f-1258-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>f91c655f-1189-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>C3L-00358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>6be95f00-ffcf-11e8-9f26-005056921935</td>\n",
       "      <td>[{'system': 'PDC', 'value': '6be95f00-ffcf-11e...</td>\n",
       "      <td>03CPTAC_UCEC_P_PNNL_20170922_B1S3_f09.psm</td>\n",
       "      <td>Peptide Spectral Matches</td>\n",
       "      <td>Text</td>\n",
       "      <td>tsv</td>\n",
       "      <td>CPTAC3-Discovery</td>\n",
       "      <td>drs://dg.4DFC:6be95f00-ffcf-11e8-9f26-00505692...</td>\n",
       "      <td>2446411</td>\n",
       "      <td>002357b9fcbb60216779c3ef40bb225b</td>\n",
       "      <td>Proteomic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>74ef895e-1271-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>f91c655f-1189-11e9-afb9-0a9c39d33490</td>\n",
       "      <td>C3L-00358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   002ff7dd-d55c-4f60-8d9a-c814443ed665   \n",
       "1   00305181-fec7-41c6-a143-f3bf4636bd5f   \n",
       "2   005ff5ca-0975-4b0b-b413-4a906701e923   \n",
       "3   007c3c99-c50f-4e45-907a-7ab78f47ff0d   \n",
       "4   00a7b64e-ffd2-11e8-abd7-005056921935   \n",
       "..                                   ...   \n",
       "95  6a95c4ba-ffb4-11e8-95a3-005056921935   \n",
       "96  6afce870-f915-11e8-90c1-005056921935   \n",
       "97  6b78fe54-ffcf-11e8-9f26-005056921935   \n",
       "98  6bcbea24-ffcf-11e8-9f26-005056921935   \n",
       "99  6be95f00-ffcf-11e8-9f26-005056921935   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '002ff7dd-d55c-4f6...   \n",
       "1   [{'system': 'GDC', 'value': '00305181-fec7-41c...   \n",
       "2   [{'system': 'GDC', 'value': '005ff5ca-0975-4b0...   \n",
       "3   [{'system': 'GDC', 'value': '007c3c99-c50f-4e4...   \n",
       "4   [{'system': 'PDC', 'value': '00a7b64e-ffd2-11e...   \n",
       "..                                                ...   \n",
       "95  [{'system': 'PDC', 'value': '6a95c4ba-ffb4-11e...   \n",
       "96  [{'system': 'PDC', 'value': '6afce870-f915-11e...   \n",
       "97  [{'system': 'PDC', 'value': '6b78fe54-ffcf-11e...   \n",
       "98  [{'system': 'PDC', 'value': '6bcbea24-ffcf-11e...   \n",
       "99  [{'system': 'PDC', 'value': '6be95f00-ffcf-11e...   \n",
       "\n",
       "                                                label  \\\n",
       "0   9e4a57ec-ca07-4843-b77f-eb5504a817c4.rna_seq.c...   \n",
       "1   39550f2d-268b-47a8-996c-7638f07de1dd.wxs.muse....   \n",
       "2   f059db9c-510f-4dc0-bf36-6038149bf06a.wxs.somat...   \n",
       "3   TCGA_UCS.7ea16309-04b4-41ff-93aa-dc3605be9fc7....   \n",
       "4       15CPTAC_UCEC_P_PNNL_20180503_B4S3_f08.mzid.gz   \n",
       "..                                                ...   \n",
       "95      07CPTAC_UCEC_W_PNNL_20170922_B2S3_f18.mzML.gz   \n",
       "96          10CPTAC_UCEC_P_PNNL_20180222_B3S2_f04.raw   \n",
       "97          03CPTAC_UCEC_P_PNNL_20170922_B1S3_f05.psm   \n",
       "98          03CPTAC_UCEC_P_PNNL_20170922_B1S3_f08.psm   \n",
       "99          03CPTAC_UCEC_P_PNNL_20170922_B1S3_f09.psm   \n",
       "\n",
       "                  data_category                    data_type      file_format  \\\n",
       "0              Sequencing Reads                Aligned Reads              BAM   \n",
       "1   Simple Nucleotide Variation  Raw Simple Somatic Mutation              VCF   \n",
       "2   Simple Nucleotide Variation  Raw Simple Somatic Mutation              VCF   \n",
       "3   Simple Nucleotide Variation   Annotated Somatic Mutation              VCF   \n",
       "4      Peptide Spectral Matches                Open Standard        mzIdentML   \n",
       "..                          ...                          ...              ...   \n",
       "95       Processed Mass Spectra                Open Standard             mzML   \n",
       "96             Raw Mass Spectra                  Proprietary  vendor-specific   \n",
       "97     Peptide Spectral Matches                         Text              tsv   \n",
       "98     Peptide Spectral Matches                         Text              tsv   \n",
       "99     Peptide Spectral Matches                         Text              tsv   \n",
       "\n",
       "   associated_project                                            drs_uri  \\\n",
       "0           TCGA-CESC  drs://dg.4DFC:002ff7dd-d55c-4f60-8d9a-c814443e...   \n",
       "1           TCGA-UCEC  drs://dg.4DFC:00305181-fec7-41c6-a143-f3bf4636...   \n",
       "2           TCGA-CESC  drs://dg.4DFC:005ff5ca-0975-4b0b-b413-4a906701...   \n",
       "3            TCGA-UCS  drs://dg.4DFC:007c3c99-c50f-4e45-907a-7ab78f47...   \n",
       "4    CPTAC3-Discovery  drs://dg.4DFC:00a7b64e-ffd2-11e8-abd7-00505692...   \n",
       "..                ...                                                ...   \n",
       "95   CPTAC3-Discovery  drs://dg.4DFC:6a95c4ba-ffb4-11e8-95a3-00505692...   \n",
       "96   CPTAC3-Discovery  drs://dg.4DFC:6afce870-f915-11e8-90c1-00505692...   \n",
       "97   CPTAC3-Discovery  drs://dg.4DFC:6b78fe54-ffcf-11e8-9f26-00505692...   \n",
       "98   CPTAC3-Discovery  drs://dg.4DFC:6bcbea24-ffcf-11e8-9f26-00505692...   \n",
       "99   CPTAC3-Discovery  drs://dg.4DFC:6be95f00-ffcf-11e8-9f26-00505692...   \n",
       "\n",
       "    byte_size                          checksum data_modality  \\\n",
       "0    96671613  6a4989df92ba58fe6558e7f6d0d5eb0f       Genomic   \n",
       "1       25035  42b1a48e34690de6cd6851ca753cff9d       Genomic   \n",
       "2      268168  4174aa64898dfbdf2a73634c73b77d01       Genomic   \n",
       "3      229002  8d2de6c206adef0e1236a1008d9234f9       Genomic   \n",
       "4     5832060  194e198498ab1c3ff18984bff2ef28a9     Proteomic   \n",
       "..        ...                               ...           ...   \n",
       "95  231084046  4eccdcd78267609a2e6e0cdd6d1f075f     Proteomic   \n",
       "96  806180410  625413aa5fece36dc7fa5375f92580ec     Proteomic   \n",
       "97    2957979  71d54336cc42e41228484d9399d22f59     Proteomic   \n",
       "98    2496941  bb2eb73c4d58f988538ca5b6b8453cfb     Proteomic   \n",
       "99    2446411  002357b9fcbb60216779c3ef40bb225b     Proteomic   \n",
       "\n",
       "   imaging_modality dbgap_accession_number imaging_series  \\\n",
       "0              None                   None           None   \n",
       "1              None                   None           None   \n",
       "2              None                   None           None   \n",
       "3              None                   None           None   \n",
       "4              None                   None           None   \n",
       "..              ...                    ...            ...   \n",
       "95             None                   None           None   \n",
       "96             None                   None           None   \n",
       "97             None                   None           None   \n",
       "98             None                   None           None   \n",
       "99             None                   None           None   \n",
       "\n",
       "             researchsubject_specimen_id  \\\n",
       "0   a36de57d-2315-4bc3-93bf-69ec4f51063b   \n",
       "1   987af661-a8b7-47b0-800f-a0b6a3ef0096   \n",
       "2   8a53198f-ee29-4cca-8cb2-9944790bcc53   \n",
       "3   e8dc4a59-fdfb-40d0-afc9-f7f21eb0f339   \n",
       "4   61c4d96e-1259-11e9-afb9-0a9c39d33490   \n",
       "..                                   ...   \n",
       "95  c8a6ef50-1259-11e9-afb9-0a9c39d33490   \n",
       "96  d799fbea-1271-11e9-afb9-0a9c39d33490   \n",
       "97  39743eb5-1259-11e9-afb9-0a9c39d33490   \n",
       "98  bb56d00f-1258-11e9-afb9-0a9c39d33490   \n",
       "99  74ef895e-1271-11e9-afb9-0a9c39d33490   \n",
       "\n",
       "                      researchsubject_id    subject_id  \n",
       "0   0bd52d75-5113-4ef1-bc75-509f59eaef2b  TCGA-JW-A5VG  \n",
       "1   e31db89c-15ac-45a4-9788-afb339b0f3bc  TCGA-AX-A3G6  \n",
       "2   e15d92a1-6925-4b58-a494-d6f665f67b83  TCGA-LP-A4AU  \n",
       "3   a1206473-cf9d-4bca-97c9-67f35817b806  TCGA-N9-A4Q1  \n",
       "4   7ef278e9-118a-11e9-afb9-0a9c39d33490     C3L-01925  \n",
       "..                                   ...           ...  \n",
       "95  bf1f593d-118a-11e9-afb9-0a9c39d33490     C3N-01217  \n",
       "96  c289029e-118a-11e9-afb9-0a9c39d33490     C3N-01267  \n",
       "97  6b043589-118a-11e9-afb9-0a9c39d33490     C3L-01256  \n",
       "98  f91c655f-1189-11e9-afb9-0a9c39d33490     C3L-00358  \n",
       "99  f91c655f-1189-11e9-afb9-0a9c39d33490     C3L-00358  \n",
       "\n",
       "[100 rows x 17 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.file.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cf9f2d3",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>label:</b> Short name or abbreviation for dataset. Maps to rdfs:label.</li>\n",
    "  <li><b>data_catagory:</b> Broad categorization of the contents of the data file.</li>\n",
    "  <li><b>data_type:</b> Specific content type of the data file.</li>\n",
    "  <li><b>file_format:</b> Format of the data files.</li>\n",
    "  <li><b>associated_project:</b> A reference to the Project(s) of which this ResearchSubject is a member. The associated_project may be embedded using the ref definition or may be a reference to the id for the Project - or a URI expressed as a string to an existing entity.</li>\n",
    "  <li><b>drs_uri:</b> A string of characters used to identify a resource on the Data Repo Service(DRS). Can be used to retreive this specific file from a server.</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes. Maps to dcat:byteSize.</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file. A digit representing the sum of the correct digits in a piece of stored or transmitted digital data, against which later comparisons can be made to detect errors in the data.</li>\n",
    "  <li><b>data_modality:</b> Data modality describes the biological nature of the information gathered as the result of an Activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\".</li>\n",
    "  <li><b>imaging_modality:</b> An imaging modality describes the imaging equipment and/or method used to acquire certain structural or functional information about the body. These include but are not limited to computed tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.</li>\n",
    "  <li><b>dbgap_accession_number:</b> The dbgap accession number for the project.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6aadbe",
   "metadata": {},
   "source": [
    "## Working with Results (pagination)\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia wants to save these results to use for the future. Since the preview dataframes only show the first 100 results of each search, she uses the `paginator` function to get all the data from the subject and researchsubject endpoints into their own dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c2cec2bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3540 ms\n"
     ]
    }
   ],
   "source": [
    "researchsubs = NoAdenoData.researchsubject.run()\n",
    "rsdf = pd.DataFrame()\n",
    "for i in researchsubs.paginator(to_df=True):\n",
    "    rsdf = pd.concat([rsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a1258057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 3538 ms\n"
     ]
    }
   ],
   "source": [
    "subs = NoAdenoData.subject.run()\n",
    "subsdf = pd.DataFrame()\n",
    "for i in subs.paginator(to_df=True):\n",
    "    subsdf = pd.concat([subsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "04cd73df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0609d8ac-fc18-4e07-b5fb-12b372151dcf</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0609d8ac-fc18-4e0...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-011667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1d2937e8-6104-4330-b7dc-7bcd79dac927</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1d2937e8-6104-433...</td>\n",
       "      <td>TCGA-UCS</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>TCGA-N5-A4RN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2639fc46-2bf7-4050-b5dd-c4d8beb89d2e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2639fc46-2bf7-405...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0003476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>30a4fa5d-98f4-4592-868f-2beb443eb82d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '30a4fa5d-98f4-459...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD13447</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30ffb337-c844-493f-97d6-7aea157d1b5e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '30ffb337-c844-493...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>da85e3db-253d-4331-bf09-b4228d6c301a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'da85e3db-253d-433...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>db7036b1-1427-49fc-9381-2df96ee45faf</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'db7036b1-1427-49f...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Trophoblastic neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-007248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>ed12e8d3-570b-4f63-b236-c807c523c7d0</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ed12e8d3-570b-4f6...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Trophoblastic neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-000252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>ede93186-676f-4ebc-a750-60607929b0d1</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ede93186-676f-4eb...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0015430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>f5afc8b8-5f01-4137-b333-6f3712bf4682</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f5afc8b8-5f01-413...</td>\n",
       "      <td>GENIE-UHN</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-UHN-365307</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3197 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   0609d8ac-fc18-4e07-b5fb-12b372151dcf   \n",
       "1   1d2937e8-6104-4330-b7dc-7bcd79dac927   \n",
       "2   2639fc46-2bf7-4050-b5dd-c4d8beb89d2e   \n",
       "3   30a4fa5d-98f4-4592-868f-2beb443eb82d   \n",
       "4   30ffb337-c844-493f-97d6-7aea157d1b5e   \n",
       "..                                   ...   \n",
       "92  da85e3db-253d-4331-bf09-b4228d6c301a   \n",
       "93  db7036b1-1427-49fc-9381-2df96ee45faf   \n",
       "94  ed12e8d3-570b-4f63-b236-c807c523c7d0   \n",
       "95  ede93186-676f-4ebc-a750-60607929b0d1   \n",
       "96  f5afc8b8-5f01-4137-b333-6f3712bf4682   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '0609d8ac-fc18-4e0...   \n",
       "1   [{'system': 'GDC', 'value': '1d2937e8-6104-433...   \n",
       "2   [{'system': 'GDC', 'value': '2639fc46-2bf7-405...   \n",
       "3   [{'system': 'GDC', 'value': '30a4fa5d-98f4-459...   \n",
       "4   [{'system': 'GDC', 'value': '30ffb337-c844-493...   \n",
       "..                                                ...   \n",
       "92  [{'system': 'GDC', 'value': 'da85e3db-253d-433...   \n",
       "93  [{'system': 'GDC', 'value': 'db7036b1-1427-49f...   \n",
       "94  [{'system': 'GDC', 'value': 'ed12e8d3-570b-4f6...   \n",
       "95  [{'system': 'GDC', 'value': 'ede93186-676f-4eb...   \n",
       "96  [{'system': 'GDC', 'value': 'f5afc8b8-5f01-413...   \n",
       "\n",
       "   member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                  GENIE-DFCI   Soft Tissue Tumors and Sarcomas, NOS   \n",
       "1                    TCGA-UCS    Complex Mixed and Stromal Neoplasms   \n",
       "2                   GENIE-MSK                    Myomatous Neoplasms   \n",
       "3                       FM-AD  Cystic, Mucinous and Serous Neoplasms   \n",
       "4               CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "..                        ...                                    ...   \n",
       "92              CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "93                 GENIE-DFCI                Trophoblastic neoplasms   \n",
       "94                 GENIE-DFCI                Trophoblastic neoplasms   \n",
       "95                  GENIE-MSK  Cystic, Mucinous and Serous Neoplasms   \n",
       "96                  GENIE-UHN                    Myomatous Neoplasms   \n",
       "\n",
       "   primary_diagnosis_site           subject_id  \n",
       "0             Uterus, NOS    GENIE-DFCI-011667  \n",
       "1             Uterus, NOS         TCGA-N5-A4RN  \n",
       "2             Uterus, NOS  GENIE-MSK-P-0003476  \n",
       "3             Uterus, NOS              AD13447  \n",
       "4            Cervix uteri    HTMCP-03-06-02240  \n",
       "..                    ...                  ...  \n",
       "92           Cervix uteri    HTMCP-03-06-02210  \n",
       "93            Uterus, NOS    GENIE-DFCI-007248  \n",
       "94            Uterus, NOS    GENIE-DFCI-000252  \n",
       "95            Uterus, NOS  GENIE-MSK-P-0015430  \n",
       "96            Uterus, NOS     GENIE-UHN-365307  \n",
       "\n",
       "[3197 rows x 6 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rsdf # view the researchsubject dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "92a6f811",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AD15141</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD15141'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AD290</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD290'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AD4252</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD4252'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AD9808</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD9808'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C3L-02382</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'C3L-02382'}]</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[cptac_ucec]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-D1-A17K</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-D1-A17K'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-27283.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TCGA-EK-A2R8</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2R8'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17692.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TCGA-N6-A4VC</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-N6-A4VC'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-30643.0</td>\n",
       "      <td>[TCGA-UCS, tcga_ucs]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>597.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TCGA-N7-A4Y5</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-N7-A4Y5'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-32058.0</td>\n",
       "      <td>[TCGA-UCS, tcga_ucs]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>8.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TCGA-VS-A8EJ</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A8EJ'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-22000.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>607.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2609 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              id                                         identifier  \\\n",
       "0        AD15141            [{'system': 'GDC', 'value': 'AD15141'}]   \n",
       "1          AD290              [{'system': 'GDC', 'value': 'AD290'}]   \n",
       "2         AD4252             [{'system': 'GDC', 'value': 'AD4252'}]   \n",
       "3         AD9808             [{'system': 'GDC', 'value': 'AD9808'}]   \n",
       "4      C3L-02382          [{'system': 'IDC', 'value': 'C3L-02382'}]   \n",
       "..           ...                                                ...   \n",
       "4   TCGA-D1-A17K  [{'system': 'GDC', 'value': 'TCGA-D1-A17K'}, {...   \n",
       "5   TCGA-EK-A2R8  [{'system': 'GDC', 'value': 'TCGA-EK-A2R8'}, {...   \n",
       "6   TCGA-N6-A4VC  [{'system': 'GDC', 'value': 'TCGA-N6-A4VC'}, {...   \n",
       "7   TCGA-N7-A4Y5  [{'system': 'GDC', 'value': 'TCGA-N7-A4Y5'}, {...   \n",
       "8   TCGA-VS-A8EJ  [{'system': 'GDC', 'value': 'TCGA-VS-A8EJ'}, {...   \n",
       "\n",
       "         species     sex          race               ethnicity  days_to_birth  \\\n",
       "0   homo sapiens  female  not reported            not reported            NaN   \n",
       "1   homo sapiens  female  not reported            not reported            NaN   \n",
       "2   homo sapiens  female  not reported            not reported            NaN   \n",
       "3   homo sapiens  female  not reported            not reported            NaN   \n",
       "4   Homo sapiens    None          None                    None            NaN   \n",
       "..           ...     ...           ...                     ...            ...   \n",
       "4   homo sapiens  female         white  not hispanic or latino       -27283.0   \n",
       "5   homo sapiens  female         white  not hispanic or latino       -17692.0   \n",
       "6   homo sapiens  female         white  not hispanic or latino       -30643.0   \n",
       "7   homo sapiens  female         white            not reported       -32058.0   \n",
       "8   homo sapiens  female         white            not reported       -22000.0   \n",
       "\n",
       "   subject_associated_project  vital_status  days_to_death cause_of_death  \n",
       "0                     [FM-AD]  Not Reported            NaN           None  \n",
       "1                     [FM-AD]  Not Reported            NaN           None  \n",
       "2                     [FM-AD]  Not Reported            NaN           None  \n",
       "3                     [FM-AD]  Not Reported            NaN           None  \n",
       "4                [cptac_ucec]          None            NaN           None  \n",
       "..                        ...           ...            ...            ...  \n",
       "4      [TCGA-UCEC, tcga_ucec]         Alive            NaN           None  \n",
       "5      [TCGA-CESC, tcga_cesc]         Alive            NaN           None  \n",
       "6        [TCGA-UCS, tcga_ucs]          Dead          597.0           None  \n",
       "7        [TCGA-UCS, tcga_ucs]          Dead            8.0           None  \n",
       "8      [TCGA-CESC, tcga_cesc]          Dead          607.0           None  \n",
       "\n",
       "[2609 rows x 11 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subsdf # view the subject dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75bcbe86",
   "metadata": {},
   "source": [
    "## Merging Results across Endpoints\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Then Julia uses the `id` fields in each result to merge them together into one big dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9b7a3383",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id_x</th>\n",
       "      <th>identifier_x</th>\n",
       "      <th>member_of_research_project</th>\n",
       "      <th>primary_diagnosis_condition</th>\n",
       "      <th>primary_diagnosis_site</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>id_y</th>\n",
       "      <th>identifier_y</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0609d8ac-fc18-4e07-b5fb-12b372151dcf</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0609d8ac-fc18-4e0...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Soft Tissue Tumors and Sarcomas, NOS</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-011667</td>\n",
       "      <td>GENIE-DFCI-011667</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-011667'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23010.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1d2937e8-6104-4330-b7dc-7bcd79dac927</td>\n",
       "      <td>[{'system': 'GDC', 'value': '1d2937e8-6104-433...</td>\n",
       "      <td>TCGA-UCS</td>\n",
       "      <td>Complex Mixed and Stromal Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>TCGA-N5-A4RN</td>\n",
       "      <td>TCGA-N5-A4RN</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-N5-A4RN'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-26054.0</td>\n",
       "      <td>[TCGA-UCS, tcga_ucs]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-N5-A4RN__tcga_ucs</td>\n",
       "      <td>[{'system': 'IDC', 'value': 'TCGA-N5-A4RN__tcg...</td>\n",
       "      <td>tcga_ucs</td>\n",
       "      <td>None</td>\n",
       "      <td>Uterus</td>\n",
       "      <td>TCGA-N5-A4RN</td>\n",
       "      <td>TCGA-N5-A4RN</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-N5-A4RN'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-26054.0</td>\n",
       "      <td>[TCGA-UCS, tcga_ucs]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2639fc46-2bf7-4050-b5dd-c4d8beb89d2e</td>\n",
       "      <td>[{'system': 'GDC', 'value': '2639fc46-2bf7-405...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0003476</td>\n",
       "      <td>GENIE-MSK-P-0003476</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00034...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-16071.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>30a4fa5d-98f4-4592-868f-2beb443eb82d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '30a4fa5d-98f4-459...</td>\n",
       "      <td>FM-AD</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>AD13447</td>\n",
       "      <td>AD13447</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'AD13447'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[FM-AD]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3192</th>\n",
       "      <td>da85e3db-253d-4331-bf09-b4228d6c301a</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'da85e3db-253d-433...</td>\n",
       "      <td>CGCI-HTMCP-CC</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Cervix uteri</td>\n",
       "      <td>HTMCP-03-06-02210</td>\n",
       "      <td>HTMCP-03-06-02210</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'HTMCP-03-06-02210'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>Unknown</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[CGCI-HTMCP-CC]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>183.0</td>\n",
       "      <td>Unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3193</th>\n",
       "      <td>db7036b1-1427-49fc-9381-2df96ee45faf</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'db7036b1-1427-49f...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Trophoblastic neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-007248</td>\n",
       "      <td>GENIE-DFCI-007248</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-007248'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-15340.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3194</th>\n",
       "      <td>ed12e8d3-570b-4f63-b236-c807c523c7d0</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ed12e8d3-570b-4f6...</td>\n",
       "      <td>GENIE-DFCI</td>\n",
       "      <td>Trophoblastic neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-DFCI-000252</td>\n",
       "      <td>GENIE-DFCI-000252</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-DFCI-000252'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-13879.0</td>\n",
       "      <td>[GENIE-DFCI]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3195</th>\n",
       "      <td>ede93186-676f-4ebc-a750-60607929b0d1</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'ede93186-676f-4eb...</td>\n",
       "      <td>GENIE-MSK</td>\n",
       "      <td>Cystic, Mucinous and Serous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-MSK-P-0015430</td>\n",
       "      <td>GENIE-MSK-P-0015430</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-MSK-P-00154...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-30681.0</td>\n",
       "      <td>[GENIE-MSK]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3196</th>\n",
       "      <td>f5afc8b8-5f01-4137-b333-6f3712bf4682</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'f5afc8b8-5f01-413...</td>\n",
       "      <td>GENIE-UHN</td>\n",
       "      <td>Myomatous Neoplasms</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "      <td>GENIE-UHN-365307</td>\n",
       "      <td>GENIE-UHN-365307</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'GENIE-UHN-365307'}]</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not allowed to collect</td>\n",
       "      <td>-24837.0</td>\n",
       "      <td>[GENIE-UHN]</td>\n",
       "      <td>Not Reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3197 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id_x  \\\n",
       "0     0609d8ac-fc18-4e07-b5fb-12b372151dcf   \n",
       "1     1d2937e8-6104-4330-b7dc-7bcd79dac927   \n",
       "2                   TCGA-N5-A4RN__tcga_ucs   \n",
       "3     2639fc46-2bf7-4050-b5dd-c4d8beb89d2e   \n",
       "4     30a4fa5d-98f4-4592-868f-2beb443eb82d   \n",
       "...                                    ...   \n",
       "3192  da85e3db-253d-4331-bf09-b4228d6c301a   \n",
       "3193  db7036b1-1427-49fc-9381-2df96ee45faf   \n",
       "3194  ed12e8d3-570b-4f63-b236-c807c523c7d0   \n",
       "3195  ede93186-676f-4ebc-a750-60607929b0d1   \n",
       "3196  f5afc8b8-5f01-4137-b333-6f3712bf4682   \n",
       "\n",
       "                                           identifier_x  \\\n",
       "0     [{'system': 'GDC', 'value': '0609d8ac-fc18-4e0...   \n",
       "1     [{'system': 'GDC', 'value': '1d2937e8-6104-433...   \n",
       "2     [{'system': 'IDC', 'value': 'TCGA-N5-A4RN__tcg...   \n",
       "3     [{'system': 'GDC', 'value': '2639fc46-2bf7-405...   \n",
       "4     [{'system': 'GDC', 'value': '30a4fa5d-98f4-459...   \n",
       "...                                                 ...   \n",
       "3192  [{'system': 'GDC', 'value': 'da85e3db-253d-433...   \n",
       "3193  [{'system': 'GDC', 'value': 'db7036b1-1427-49f...   \n",
       "3194  [{'system': 'GDC', 'value': 'ed12e8d3-570b-4f6...   \n",
       "3195  [{'system': 'GDC', 'value': 'ede93186-676f-4eb...   \n",
       "3196  [{'system': 'GDC', 'value': 'f5afc8b8-5f01-413...   \n",
       "\n",
       "     member_of_research_project            primary_diagnosis_condition  \\\n",
       "0                    GENIE-DFCI   Soft Tissue Tumors and Sarcomas, NOS   \n",
       "1                      TCGA-UCS    Complex Mixed and Stromal Neoplasms   \n",
       "2                      tcga_ucs                                   None   \n",
       "3                     GENIE-MSK                    Myomatous Neoplasms   \n",
       "4                         FM-AD  Cystic, Mucinous and Serous Neoplasms   \n",
       "...                         ...                                    ...   \n",
       "3192              CGCI-HTMCP-CC                Squamous Cell Neoplasms   \n",
       "3193                 GENIE-DFCI                Trophoblastic neoplasms   \n",
       "3194                 GENIE-DFCI                Trophoblastic neoplasms   \n",
       "3195                  GENIE-MSK  Cystic, Mucinous and Serous Neoplasms   \n",
       "3196                  GENIE-UHN                    Myomatous Neoplasms   \n",
       "\n",
       "     primary_diagnosis_site           subject_id                 id_y  \\\n",
       "0               Uterus, NOS    GENIE-DFCI-011667    GENIE-DFCI-011667   \n",
       "1               Uterus, NOS         TCGA-N5-A4RN         TCGA-N5-A4RN   \n",
       "2                    Uterus         TCGA-N5-A4RN         TCGA-N5-A4RN   \n",
       "3               Uterus, NOS  GENIE-MSK-P-0003476  GENIE-MSK-P-0003476   \n",
       "4               Uterus, NOS              AD13447              AD13447   \n",
       "...                     ...                  ...                  ...   \n",
       "3192           Cervix uteri    HTMCP-03-06-02210    HTMCP-03-06-02210   \n",
       "3193            Uterus, NOS    GENIE-DFCI-007248    GENIE-DFCI-007248   \n",
       "3194            Uterus, NOS    GENIE-DFCI-000252    GENIE-DFCI-000252   \n",
       "3195            Uterus, NOS  GENIE-MSK-P-0015430  GENIE-MSK-P-0015430   \n",
       "3196            Uterus, NOS     GENIE-UHN-365307     GENIE-UHN-365307   \n",
       "\n",
       "                                           identifier_y       species     sex  \\\n",
       "0     [{'system': 'GDC', 'value': 'GENIE-DFCI-011667'}]  homo sapiens  female   \n",
       "1     [{'system': 'GDC', 'value': 'TCGA-N5-A4RN'}, {...  homo sapiens  female   \n",
       "2     [{'system': 'GDC', 'value': 'TCGA-N5-A4RN'}, {...  homo sapiens  female   \n",
       "3     [{'system': 'GDC', 'value': 'GENIE-MSK-P-00034...  homo sapiens  female   \n",
       "4               [{'system': 'GDC', 'value': 'AD13447'}]  homo sapiens  female   \n",
       "...                                                 ...           ...     ...   \n",
       "3192  [{'system': 'GDC', 'value': 'HTMCP-03-06-02210'}]  homo sapiens  female   \n",
       "3193  [{'system': 'GDC', 'value': 'GENIE-DFCI-007248'}]  homo sapiens  female   \n",
       "3194  [{'system': 'GDC', 'value': 'GENIE-DFCI-000252'}]  homo sapiens  female   \n",
       "3195  [{'system': 'GDC', 'value': 'GENIE-MSK-P-00154...  homo sapiens  female   \n",
       "3196   [{'system': 'GDC', 'value': 'GENIE-UHN-365307'}]  homo sapiens  female   \n",
       "\n",
       "                           race               ethnicity  days_to_birth  \\\n",
       "0                         white  not hispanic or latino       -23010.0   \n",
       "1                         white  not hispanic or latino       -26054.0   \n",
       "2                         white  not hispanic or latino       -26054.0   \n",
       "3                         white  not hispanic or latino       -16071.0   \n",
       "4                  not reported            not reported            NaN   \n",
       "...                         ...                     ...            ...   \n",
       "3192  black or african american                 Unknown            NaN   \n",
       "3193                      white  not hispanic or latino       -15340.0   \n",
       "3194                      white  not hispanic or latino       -13879.0   \n",
       "3195                      white  not hispanic or latino       -30681.0   \n",
       "3196                      white  not allowed to collect       -24837.0   \n",
       "\n",
       "     subject_associated_project  vital_status  days_to_death cause_of_death  \n",
       "0                  [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "1          [TCGA-UCS, tcga_ucs]         Alive            NaN           None  \n",
       "2          [TCGA-UCS, tcga_ucs]         Alive            NaN           None  \n",
       "3                   [GENIE-MSK]  Not Reported            NaN           None  \n",
       "4                       [FM-AD]  Not Reported            NaN           None  \n",
       "...                         ...           ...            ...            ...  \n",
       "3192            [CGCI-HTMCP-CC]          Dead          183.0        Unknown  \n",
       "3193               [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "3194               [GENIE-DFCI]  Not Reported            NaN           None  \n",
       "3195                [GENIE-MSK]  Not Reported            NaN           None  \n",
       "3196                [GENIE-UHN]  Not Reported            NaN           None  \n",
       "\n",
       "[3197 rows x 17 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allmetadata = pd.merge(rsdf,\n",
    "                subsdf,\n",
    "                left_on=\"subject_id\",\n",
    "                right_on='id')\n",
    "\n",
    "allmetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "024da831",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "And saves it out to a csv so she can browse it with Excel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b6628de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "allmetadata.to_csv(\"allmetadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246644d3",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia knows from her subject count summary that there are more than 200,000 files associated with her subjects, which is likely far more than she needs. To help her decide what files she wants, Julia uses endpoint chaining to get summary information about the files that are assigned to researchsubjects for her search criteria:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ae1ae079",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 13151 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">   total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">267283</span>   \n",
       "</pre>\n"
      ],
      "text/plain": [
       "   total : \u001b[1;36m267283\u001b[0m   \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_736fe_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_736fe_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_736fe_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_736fe_row0_col0\" class=\"data row0 col0\" >IDC</td>\n",
       "      <td id=\"T_736fe_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_736fe_row1_col0\" class=\"data row1 col0\" >GDC</td>\n",
       "      <td id=\"T_736fe_row1_col1\" class=\"data row1 col1\" >294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_736fe_row2_col0\" class=\"data row2 col0\" >PDC</td>\n",
       "      <td id=\"T_736fe_row2_col1\" class=\"data row2 col1\" >2560</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_74c1b_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_74c1b_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_74c1b_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row0_col0\" class=\"data row0 col0\" >Imaging</td>\n",
       "      <td id=\"T_74c1b_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row1_col0\" class=\"data row1 col0\" >Raw Mass Spectra</td>\n",
       "      <td id=\"T_74c1b_row1_col1\" class=\"data row1 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row2_col0\" class=\"data row2 col0\" >Peptide Spectral Matches</td>\n",
       "      <td id=\"T_74c1b_row2_col1\" class=\"data row2 col1\" >1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row3_col0\" class=\"data row3 col0\" >Processed Mass Spectra</td>\n",
       "      <td id=\"T_74c1b_row3_col1\" class=\"data row3 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row4_col0\" class=\"data row4 col0\" >Simple Nucleotide Variation</td>\n",
       "      <td id=\"T_74c1b_row4_col1\" class=\"data row4 col1\" >114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row5_col0\" class=\"data row5 col0\" >Somatic Structural Variation</td>\n",
       "      <td id=\"T_74c1b_row5_col1\" class=\"data row5 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row6_col0\" class=\"data row6 col0\" >Copy Number Variation</td>\n",
       "      <td id=\"T_74c1b_row6_col1\" class=\"data row6 col1\" >37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row7_col0\" class=\"data row7 col0\" >Transcriptome Profiling</td>\n",
       "      <td id=\"T_74c1b_row7_col1\" class=\"data row7 col1\" >31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row8_col0\" class=\"data row8 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_74c1b_row8_col1\" class=\"data row8 col1\" >17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row9_col0\" class=\"data row9 col0\" >Structural Variation</td>\n",
       "      <td id=\"T_74c1b_row9_col1\" class=\"data row9 col1\" >29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row10_col0\" class=\"data row10 col0\" >Clinical</td>\n",
       "      <td id=\"T_74c1b_row10_col1\" class=\"data row10 col1\" >3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row11_col0\" class=\"data row11 col0\" >DNA Methylation</td>\n",
       "      <td id=\"T_74c1b_row11_col1\" class=\"data row11 col1\" >24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_74c1b_row12_col0\" class=\"data row12 col0\" >Sequencing Reads</td>\n",
       "      <td id=\"T_74c1b_row12_col1\" class=\"data row12 col1\" >35</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_4bfb1_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_4bfb1_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_4bfb1_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row0_col0\" class=\"data row0 col0\" >DICOM</td>\n",
       "      <td id=\"T_4bfb1_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row1_col0\" class=\"data row1 col0\" >TSV</td>\n",
       "      <td id=\"T_4bfb1_row1_col1\" class=\"data row1 col1\" >39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row2_col0\" class=\"data row2 col0\" >tsv</td>\n",
       "      <td id=\"T_4bfb1_row2_col1\" class=\"data row2 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row3_col0\" class=\"data row3 col0\" >vendor-specific</td>\n",
       "      <td id=\"T_4bfb1_row3_col1\" class=\"data row3 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row4_col0\" class=\"data row4 col0\" >BAM</td>\n",
       "      <td id=\"T_4bfb1_row4_col1\" class=\"data row4 col1\" >35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row5_col0\" class=\"data row5 col0\" >mzML</td>\n",
       "      <td id=\"T_4bfb1_row5_col1\" class=\"data row5 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row6_col0\" class=\"data row6 col0\" >BEDPE</td>\n",
       "      <td id=\"T_4bfb1_row6_col1\" class=\"data row6 col1\" >19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row7_col0\" class=\"data row7 col0\" >mzIdentML</td>\n",
       "      <td id=\"T_4bfb1_row7_col1\" class=\"data row7 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row8_col0\" class=\"data row8 col0\" >VCF</td>\n",
       "      <td id=\"T_4bfb1_row8_col1\" class=\"data row8 col1\" >61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row9_col0\" class=\"data row9 col0\" >TXT</td>\n",
       "      <td id=\"T_4bfb1_row9_col1\" class=\"data row9 col1\" >50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row10_col0\" class=\"data row10 col0\" >MAF</td>\n",
       "      <td id=\"T_4bfb1_row10_col1\" class=\"data row10 col1\" >54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row11_col0\" class=\"data row11 col0\" >IDAT</td>\n",
       "      <td id=\"T_4bfb1_row11_col1\" class=\"data row11 col1\" >16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row12_col0\" class=\"data row12 col0\" >SVS</td>\n",
       "      <td id=\"T_4bfb1_row12_col1\" class=\"data row12 col1\" >8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row13_col0\" class=\"data row13 col0\" >BCR Auxiliary XML</td>\n",
       "      <td id=\"T_4bfb1_row13_col1\" class=\"data row13 col1\" >1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row14_col0\" class=\"data row14 col0\" >BCR XML</td>\n",
       "      <td id=\"T_4bfb1_row14_col1\" class=\"data row14 col1\" >6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row15_col0\" class=\"data row15 col0\" >BCR SSF XML</td>\n",
       "      <td id=\"T_4bfb1_row15_col1\" class=\"data row15 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_4bfb1_row16_col0\" class=\"data row16 col0\" >BCR PPS XML</td>\n",
       "      <td id=\"T_4bfb1_row16_col1\" class=\"data row16 col1\" >1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_dd7ee_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_dd7ee_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_dd7ee_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row0_col0\" class=\"data row0 col0\" >None</td>\n",
       "      <td id=\"T_dd7ee_row0_col1\" class=\"data row0 col1\" >264429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row1_col0\" class=\"data row1 col0\" >Text</td>\n",
       "      <td id=\"T_dd7ee_row1_col1\" class=\"data row1 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row2_col0\" class=\"data row2 col0\" >Open Standard</td>\n",
       "      <td id=\"T_dd7ee_row2_col1\" class=\"data row2 col1\" >1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row3_col0\" class=\"data row3 col0\" >Proprietary</td>\n",
       "      <td id=\"T_dd7ee_row3_col1\" class=\"data row3 col1\" >640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row4_col0\" class=\"data row4 col0\" >Aligned Reads</td>\n",
       "      <td id=\"T_dd7ee_row4_col1\" class=\"data row4 col1\" >35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row5_col0\" class=\"data row5 col0\" >Copy Number Segment</td>\n",
       "      <td id=\"T_dd7ee_row5_col1\" class=\"data row5 col1\" >11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row6_col0\" class=\"data row6 col0\" >Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_dd7ee_row6_col1\" class=\"data row6 col1\" >55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row7_col0\" class=\"data row7 col0\" >Gene Level Copy Number Scores</td>\n",
       "      <td id=\"T_dd7ee_row7_col1\" class=\"data row7 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row8_col0\" class=\"data row8 col0\" >Gene Level Copy Number</td>\n",
       "      <td id=\"T_dd7ee_row8_col1\" class=\"data row8 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row9_col0\" class=\"data row9 col0\" >Raw Simple Somatic Mutation</td>\n",
       "      <td id=\"T_dd7ee_row9_col1\" class=\"data row9 col1\" >36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row10_col0\" class=\"data row10 col0\" >Gene Expression Quantification</td>\n",
       "      <td id=\"T_dd7ee_row10_col1\" class=\"data row10 col1\" >9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row11_col0\" class=\"data row11 col0\" >Slide Image</td>\n",
       "      <td id=\"T_dd7ee_row11_col1\" class=\"data row11 col1\" >8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row12_col0\" class=\"data row12 col0\" >Aggregated Somatic Mutation</td>\n",
       "      <td id=\"T_dd7ee_row12_col1\" class=\"data row12 col1\" >7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row13_col0\" class=\"data row13 col0\" >miRNA Expression Quantification</td>\n",
       "      <td id=\"T_dd7ee_row13_col1\" class=\"data row13 col1\" >7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row14_col0\" class=\"data row14 col0\" >Masked Annotated Somatic Mutation</td>\n",
       "      <td id=\"T_dd7ee_row14_col1\" class=\"data row14 col1\" >11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row15_col0\" class=\"data row15 col0\" >Masked Copy Number Segment</td>\n",
       "      <td id=\"T_dd7ee_row15_col1\" class=\"data row15 col1\" >10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row16_col0\" class=\"data row16 col0\" >Transcript Fusion</td>\n",
       "      <td id=\"T_dd7ee_row16_col1\" class=\"data row16 col1\" >29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row17_col0\" class=\"data row17 col0\" >Methylation Beta Value</td>\n",
       "      <td id=\"T_dd7ee_row17_col1\" class=\"data row17 col1\" >8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row18_col0\" class=\"data row18 col0\" >Masked Somatic Mutation</td>\n",
       "      <td id=\"T_dd7ee_row18_col1\" class=\"data row18 col1\" >5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row19_col0\" class=\"data row19 col0\" >Masked Intensities</td>\n",
       "      <td id=\"T_dd7ee_row19_col1\" class=\"data row19 col1\" >16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row20_col0\" class=\"data row20 col0\" >Biospecimen Supplement</td>\n",
       "      <td id=\"T_dd7ee_row20_col1\" class=\"data row20 col1\" >9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row21_col0\" class=\"data row21 col0\" >Allele-specific Copy Number Segment</td>\n",
       "      <td id=\"T_dd7ee_row21_col1\" class=\"data row21 col1\" >6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row22_col0\" class=\"data row22 col0\" >Splice Junction Quantification</td>\n",
       "      <td id=\"T_dd7ee_row22_col1\" class=\"data row22 col1\" >11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row23_col0\" class=\"data row23 col0\" >Isoform Expression Quantification</td>\n",
       "      <td id=\"T_dd7ee_row23_col1\" class=\"data row23 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row24_col0\" class=\"data row24 col0\" >Structural Rearrangement</td>\n",
       "      <td id=\"T_dd7ee_row24_col1\" class=\"data row24 col1\" >4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td id=\"T_dd7ee_row25_col0\" class=\"data row25 col0\" >Clinical Supplement</td>\n",
       "      <td id=\"T_dd7ee_row25_col1\" class=\"data row25 col1\" >3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "NoAdenoData.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "111f47c0",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Julia decides that a good place to start would be with Slide Images. There's only 1111, so she should be able to quickly scan through them over the next few days and see if they will be useful. So she adds one more filter on her search:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d4a170b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 7349 ms\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">     total : <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">8</span>      \n",
       "</pre>\n"
      ],
      "text/plain": [
       "     total : \u001b[1;36m8\u001b[0m      \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_d9a3e_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_d9a3e_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_d9a3e_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >system</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_d9a3e_row0_col0\" class=\"data row0 col0\" >GDC</td>\n",
       "      <td id=\"T_d9a3e_row0_col1\" class=\"data row0 col1\" >8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_0715b_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_0715b_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_0715b_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_category</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_0715b_row0_col0\" class=\"data row0 col0\" >Biospecimen</td>\n",
       "      <td id=\"T_0715b_row0_col1\" class=\"data row0 col1\" >8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_8f212_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_8f212_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_8f212_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >file_format</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_8f212_row0_col0\" class=\"data row0 col0\" >SVS</td>\n",
       "      <td id=\"T_8f212_row0_col1\" class=\"data row0 col1\" >8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<style type=\"text/css\">\n",
       "#T_b72a2_ th {\n",
       "  background-color: #000066;\n",
       "  color: white;\n",
       "  text-align: left;\n",
       "}\n",
       "#T_b72a2_ td {\n",
       "  text-align: left;\n",
       "  border-bottom: 1px solid black;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_b72a2_\" style='display:inline'>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"col_heading level0 col0\" >data_type</th>\n",
       "      <th class=\"col_heading level0 col1\" >count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td id=\"T_b72a2_row0_col0\" class=\"data row0 col0\" >Slide Image</td>\n",
       "      <td id=\"T_b72a2_row0_col1\" class=\"data row0 col1\" >8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": []
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "JustSlides = Q('file.data_type = \"Slide Image\"')\n",
    "NoadenoJustSlides = NoAdenoData.AND(JustSlides)\n",
    "NoadenoJustSlides.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e385faf",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Finally, Julia uses the pagenation function again to get all the slide files, and merges her metadata file with this file information. This way she will be able to review what phenotypes each slide is associated with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f45f6503",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">Getting results from database\n",
       "\n",
       "</pre>\n"
      ],
      "text/plain": [
       "Getting results from database\n",
       "\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total execution time: 7162 ms\n"
     ]
    }
   ],
   "source": [
    "slides = NoadenoJustSlides.researchsubject.file.run()\n",
    "slidesdf = pd.DataFrame()\n",
    "for i in slides.paginator(to_df=True):\n",
    "    slidesdf = pd.concat([slidesdf, i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0d1f7a21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>identifier</th>\n",
       "      <th>label</th>\n",
       "      <th>data_category</th>\n",
       "      <th>data_type</th>\n",
       "      <th>file_format</th>\n",
       "      <th>associated_project</th>\n",
       "      <th>drs_uri</th>\n",
       "      <th>byte_size</th>\n",
       "      <th>checksum</th>\n",
       "      <th>...</th>\n",
       "      <th>identifier_y</th>\n",
       "      <th>species</th>\n",
       "      <th>sex</th>\n",
       "      <th>race</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>subject_associated_project</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>cause_of_death</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>002a60d4-280c-44b5-a0bf-5e2f063253b2</td>\n",
       "      <td>[{'system': 'GDC', 'value': '002a60d4-280c-44b...</td>\n",
       "      <td>TCGA-EA-A6QX-01A-01-TSA.66853F14-1300-4D44-B29...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:002a60d4-280c-44b5-a0bf-5e2f0632...</td>\n",
       "      <td>256639621</td>\n",
       "      <td>c937af221666a0be405b665631582b32</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EA-A6QX'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-18035.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>002a60d4-280c-44b5-a0bf-5e2f063253b2</td>\n",
       "      <td>[{'system': 'GDC', 'value': '002a60d4-280c-44b...</td>\n",
       "      <td>TCGA-EA-A6QX-01A-01-TSA.66853F14-1300-4D44-B29...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:002a60d4-280c-44b5-a0bf-5e2f0632...</td>\n",
       "      <td>256639621</td>\n",
       "      <td>c937af221666a0be405b665631582b32</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EA-A6QX'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-18035.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>016cff56-c7a4-462b-8a84-12d112a90739</td>\n",
       "      <td>[{'system': 'GDC', 'value': '016cff56-c7a4-462...</td>\n",
       "      <td>TCGA-EK-A2H1-01A-01-TSA.B0EA350D-B3CA-482F-AC0...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:016cff56-c7a4-462b-8a84-12d112a9...</td>\n",
       "      <td>133766513</td>\n",
       "      <td>ec96548913e8060f75db0c867bb00336</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2H1'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-7649.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>016cff56-c7a4-462b-8a84-12d112a90739</td>\n",
       "      <td>[{'system': 'GDC', 'value': '016cff56-c7a4-462...</td>\n",
       "      <td>TCGA-EK-A2H1-01A-01-TSA.B0EA350D-B3CA-482F-AC0...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:016cff56-c7a4-462b-8a84-12d112a9...</td>\n",
       "      <td>133766513</td>\n",
       "      <td>ec96548913e8060f75db0c867bb00336</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-EK-A2H1'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-7649.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>01e9b85a-e6c7-4117-9bfd-86050f5a2ff7</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01e9b85a-e6c7-411...</td>\n",
       "      <td>TCGA-IW-A3M4-01Z-00-DX1.03BE04CC-552B-40E8-85D...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-SARC</td>\n",
       "      <td>drs://dg.4DFC:01e9b85a-e6c7-4117-9bfd-86050f5a...</td>\n",
       "      <td>2303418723</td>\n",
       "      <td>6ef35302e17fdb5f712c0ece967ad543</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-IW-A3M4'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-17721.0</td>\n",
       "      <td>[tcga_sarc, TCGA-SARC]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>00d9d6d3-07af-41f8-bf84-4f2bf200461d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00d9d6d3-07af-41f...</td>\n",
       "      <td>TCGA-AX-A2H5-11A-01-TSA.6a778f9f-2717-4bb8-83d...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:00d9d6d3-07af-41f8-bf84-4f2bf200...</td>\n",
       "      <td>56262805</td>\n",
       "      <td>58733b0552dc9dde57605e4215550b51</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AX-A2H5'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24757.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>00d9d6d3-07af-41f8-bf84-4f2bf200461d</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00d9d6d3-07af-41f...</td>\n",
       "      <td>TCGA-AX-A2H5-11A-01-TSA.6a778f9f-2717-4bb8-83d...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:00d9d6d3-07af-41f8-bf84-4f2bf200...</td>\n",
       "      <td>56262805</td>\n",
       "      <td>58733b0552dc9dde57605e4215550b51</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-AX-A2H5'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-24757.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>00af722b-dd64-4909-92fa-27116bd2d728</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00af722b-dd64-490...</td>\n",
       "      <td>TCGA-FU-A3HZ-01Z-00-DX1.1E78D8EF-B1FF-49AC-9EC...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:00af722b-dd64-4909-92fa-27116bd2...</td>\n",
       "      <td>538854312</td>\n",
       "      <td>1a799ba38252a841393ef5aab4f87764</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-FU-A3HZ'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23565.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>00af722b-dd64-4909-92fa-27116bd2d728</td>\n",
       "      <td>[{'system': 'GDC', 'value': '00af722b-dd64-490...</td>\n",
       "      <td>TCGA-FU-A3HZ-01Z-00-DX1.1E78D8EF-B1FF-49AC-9EC...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:00af722b-dd64-4909-92fa-27116bd2...</td>\n",
       "      <td>538854312</td>\n",
       "      <td>1a799ba38252a841393ef5aab4f87764</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-FU-A3HZ'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>-23565.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>02739e88-15df-4e7b-ac86-c0e2931bf713</td>\n",
       "      <td>[{'system': 'GDC', 'value': '02739e88-15df-4e7...</td>\n",
       "      <td>TCGA-VS-A9UL-01Z-00-DX1.4348D92F-0EA0-4A98-8F5...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:02739e88-15df-4e7b-ac86-c0e2931b...</td>\n",
       "      <td>341653835</td>\n",
       "      <td>0c506efa82b701b2ef0f3bba8b275db0</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A9UL'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-29033.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>442.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>02739e88-15df-4e7b-ac86-c0e2931bf713</td>\n",
       "      <td>[{'system': 'GDC', 'value': '02739e88-15df-4e7...</td>\n",
       "      <td>TCGA-VS-A9UL-01Z-00-DX1.4348D92F-0EA0-4A98-8F5...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:02739e88-15df-4e7b-ac86-c0e2931b...</td>\n",
       "      <td>341653835</td>\n",
       "      <td>0c506efa82b701b2ef0f3bba8b275db0</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A9UL'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-29033.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Dead</td>\n",
       "      <td>442.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>0261cb79-9879-480a-9626-63fd50d6daea</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0261cb79-9879-480...</td>\n",
       "      <td>TCGA-VS-A9UP-01A-01-TS1.68E82753-9B6E-4544-B91...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:0261cb79-9879-480a-9626-63fd50d6...</td>\n",
       "      <td>201483703</td>\n",
       "      <td>8318b25bce4cc03960bdbca1112c520b</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A9UP'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-15849.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>0261cb79-9879-480a-9626-63fd50d6daea</td>\n",
       "      <td>[{'system': 'GDC', 'value': '0261cb79-9879-480...</td>\n",
       "      <td>TCGA-VS-A9UP-01A-01-TS1.68E82753-9B6E-4544-B91...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-CESC</td>\n",
       "      <td>drs://dg.4DFC:0261cb79-9879-480a-9626-63fd50d6...</td>\n",
       "      <td>201483703</td>\n",
       "      <td>8318b25bce4cc03960bdbca1112c520b</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-VS-A9UP'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>not reported</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-15849.0</td>\n",
       "      <td>[TCGA-CESC, tcga_cesc]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>01f8a923-19c7-43c4-b337-fca47fda7a0c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01f8a923-19c7-43c...</td>\n",
       "      <td>TCGA-B5-A0JN-01A-01-BS1.19bbff7a-2913-4963-8bb...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:01f8a923-19c7-43c4-b337-fca47fda...</td>\n",
       "      <td>329775631</td>\n",
       "      <td>463372cea1072caab6e8b21f3408575c</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-B5-A0JN'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-30933.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>01f8a923-19c7-43c4-b337-fca47fda7a0c</td>\n",
       "      <td>[{'system': 'GDC', 'value': '01f8a923-19c7-43c...</td>\n",
       "      <td>TCGA-B5-A0JN-01A-01-BS1.19bbff7a-2913-4963-8bb...</td>\n",
       "      <td>Biospecimen</td>\n",
       "      <td>Slide Image</td>\n",
       "      <td>SVS</td>\n",
       "      <td>TCGA-UCEC</td>\n",
       "      <td>drs://dg.4DFC:01f8a923-19c7-43c4-b337-fca47fda...</td>\n",
       "      <td>329775631</td>\n",
       "      <td>463372cea1072caab6e8b21f3408575c</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'system': 'GDC', 'value': 'TCGA-B5-A0JN'}, {...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>female</td>\n",
       "      <td>black or african american</td>\n",
       "      <td>not reported</td>\n",
       "      <td>-30933.0</td>\n",
       "      <td>[TCGA-UCEC, tcga_ucec]</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      id  \\\n",
       "0   002a60d4-280c-44b5-a0bf-5e2f063253b2   \n",
       "1   002a60d4-280c-44b5-a0bf-5e2f063253b2   \n",
       "2   016cff56-c7a4-462b-8a84-12d112a90739   \n",
       "3   016cff56-c7a4-462b-8a84-12d112a90739   \n",
       "4   01e9b85a-e6c7-4117-9bfd-86050f5a2ff7   \n",
       "5   00d9d6d3-07af-41f8-bf84-4f2bf200461d   \n",
       "6   00d9d6d3-07af-41f8-bf84-4f2bf200461d   \n",
       "7   00af722b-dd64-4909-92fa-27116bd2d728   \n",
       "8   00af722b-dd64-4909-92fa-27116bd2d728   \n",
       "9   02739e88-15df-4e7b-ac86-c0e2931bf713   \n",
       "10  02739e88-15df-4e7b-ac86-c0e2931bf713   \n",
       "11  0261cb79-9879-480a-9626-63fd50d6daea   \n",
       "12  0261cb79-9879-480a-9626-63fd50d6daea   \n",
       "13  01f8a923-19c7-43c4-b337-fca47fda7a0c   \n",
       "14  01f8a923-19c7-43c4-b337-fca47fda7a0c   \n",
       "\n",
       "                                           identifier  \\\n",
       "0   [{'system': 'GDC', 'value': '002a60d4-280c-44b...   \n",
       "1   [{'system': 'GDC', 'value': '002a60d4-280c-44b...   \n",
       "2   [{'system': 'GDC', 'value': '016cff56-c7a4-462...   \n",
       "3   [{'system': 'GDC', 'value': '016cff56-c7a4-462...   \n",
       "4   [{'system': 'GDC', 'value': '01e9b85a-e6c7-411...   \n",
       "5   [{'system': 'GDC', 'value': '00d9d6d3-07af-41f...   \n",
       "6   [{'system': 'GDC', 'value': '00d9d6d3-07af-41f...   \n",
       "7   [{'system': 'GDC', 'value': '00af722b-dd64-490...   \n",
       "8   [{'system': 'GDC', 'value': '00af722b-dd64-490...   \n",
       "9   [{'system': 'GDC', 'value': '02739e88-15df-4e7...   \n",
       "10  [{'system': 'GDC', 'value': '02739e88-15df-4e7...   \n",
       "11  [{'system': 'GDC', 'value': '0261cb79-9879-480...   \n",
       "12  [{'system': 'GDC', 'value': '0261cb79-9879-480...   \n",
       "13  [{'system': 'GDC', 'value': '01f8a923-19c7-43c...   \n",
       "14  [{'system': 'GDC', 'value': '01f8a923-19c7-43c...   \n",
       "\n",
       "                                                label data_category  \\\n",
       "0   TCGA-EA-A6QX-01A-01-TSA.66853F14-1300-4D44-B29...   Biospecimen   \n",
       "1   TCGA-EA-A6QX-01A-01-TSA.66853F14-1300-4D44-B29...   Biospecimen   \n",
       "2   TCGA-EK-A2H1-01A-01-TSA.B0EA350D-B3CA-482F-AC0...   Biospecimen   \n",
       "3   TCGA-EK-A2H1-01A-01-TSA.B0EA350D-B3CA-482F-AC0...   Biospecimen   \n",
       "4   TCGA-IW-A3M4-01Z-00-DX1.03BE04CC-552B-40E8-85D...   Biospecimen   \n",
       "5   TCGA-AX-A2H5-11A-01-TSA.6a778f9f-2717-4bb8-83d...   Biospecimen   \n",
       "6   TCGA-AX-A2H5-11A-01-TSA.6a778f9f-2717-4bb8-83d...   Biospecimen   \n",
       "7   TCGA-FU-A3HZ-01Z-00-DX1.1E78D8EF-B1FF-49AC-9EC...   Biospecimen   \n",
       "8   TCGA-FU-A3HZ-01Z-00-DX1.1E78D8EF-B1FF-49AC-9EC...   Biospecimen   \n",
       "9   TCGA-VS-A9UL-01Z-00-DX1.4348D92F-0EA0-4A98-8F5...   Biospecimen   \n",
       "10  TCGA-VS-A9UL-01Z-00-DX1.4348D92F-0EA0-4A98-8F5...   Biospecimen   \n",
       "11  TCGA-VS-A9UP-01A-01-TS1.68E82753-9B6E-4544-B91...   Biospecimen   \n",
       "12  TCGA-VS-A9UP-01A-01-TS1.68E82753-9B6E-4544-B91...   Biospecimen   \n",
       "13  TCGA-B5-A0JN-01A-01-BS1.19bbff7a-2913-4963-8bb...   Biospecimen   \n",
       "14  TCGA-B5-A0JN-01A-01-BS1.19bbff7a-2913-4963-8bb...   Biospecimen   \n",
       "\n",
       "      data_type file_format associated_project  \\\n",
       "0   Slide Image         SVS          TCGA-CESC   \n",
       "1   Slide Image         SVS          TCGA-CESC   \n",
       "2   Slide Image         SVS          TCGA-CESC   \n",
       "3   Slide Image         SVS          TCGA-CESC   \n",
       "4   Slide Image         SVS          TCGA-SARC   \n",
       "5   Slide Image         SVS          TCGA-UCEC   \n",
       "6   Slide Image         SVS          TCGA-UCEC   \n",
       "7   Slide Image         SVS          TCGA-CESC   \n",
       "8   Slide Image         SVS          TCGA-CESC   \n",
       "9   Slide Image         SVS          TCGA-CESC   \n",
       "10  Slide Image         SVS          TCGA-CESC   \n",
       "11  Slide Image         SVS          TCGA-CESC   \n",
       "12  Slide Image         SVS          TCGA-CESC   \n",
       "13  Slide Image         SVS          TCGA-UCEC   \n",
       "14  Slide Image         SVS          TCGA-UCEC   \n",
       "\n",
       "                                              drs_uri   byte_size  \\\n",
       "0   drs://dg.4DFC:002a60d4-280c-44b5-a0bf-5e2f0632...   256639621   \n",
       "1   drs://dg.4DFC:002a60d4-280c-44b5-a0bf-5e2f0632...   256639621   \n",
       "2   drs://dg.4DFC:016cff56-c7a4-462b-8a84-12d112a9...   133766513   \n",
       "3   drs://dg.4DFC:016cff56-c7a4-462b-8a84-12d112a9...   133766513   \n",
       "4   drs://dg.4DFC:01e9b85a-e6c7-4117-9bfd-86050f5a...  2303418723   \n",
       "5   drs://dg.4DFC:00d9d6d3-07af-41f8-bf84-4f2bf200...    56262805   \n",
       "6   drs://dg.4DFC:00d9d6d3-07af-41f8-bf84-4f2bf200...    56262805   \n",
       "7   drs://dg.4DFC:00af722b-dd64-4909-92fa-27116bd2...   538854312   \n",
       "8   drs://dg.4DFC:00af722b-dd64-4909-92fa-27116bd2...   538854312   \n",
       "9   drs://dg.4DFC:02739e88-15df-4e7b-ac86-c0e2931b...   341653835   \n",
       "10  drs://dg.4DFC:02739e88-15df-4e7b-ac86-c0e2931b...   341653835   \n",
       "11  drs://dg.4DFC:0261cb79-9879-480a-9626-63fd50d6...   201483703   \n",
       "12  drs://dg.4DFC:0261cb79-9879-480a-9626-63fd50d6...   201483703   \n",
       "13  drs://dg.4DFC:01f8a923-19c7-43c4-b337-fca47fda...   329775631   \n",
       "14  drs://dg.4DFC:01f8a923-19c7-43c4-b337-fca47fda...   329775631   \n",
       "\n",
       "                            checksum  ...  \\\n",
       "0   c937af221666a0be405b665631582b32  ...   \n",
       "1   c937af221666a0be405b665631582b32  ...   \n",
       "2   ec96548913e8060f75db0c867bb00336  ...   \n",
       "3   ec96548913e8060f75db0c867bb00336  ...   \n",
       "4   6ef35302e17fdb5f712c0ece967ad543  ...   \n",
       "5   58733b0552dc9dde57605e4215550b51  ...   \n",
       "6   58733b0552dc9dde57605e4215550b51  ...   \n",
       "7   1a799ba38252a841393ef5aab4f87764  ...   \n",
       "8   1a799ba38252a841393ef5aab4f87764  ...   \n",
       "9   0c506efa82b701b2ef0f3bba8b275db0  ...   \n",
       "10  0c506efa82b701b2ef0f3bba8b275db0  ...   \n",
       "11  8318b25bce4cc03960bdbca1112c520b  ...   \n",
       "12  8318b25bce4cc03960bdbca1112c520b  ...   \n",
       "13  463372cea1072caab6e8b21f3408575c  ...   \n",
       "14  463372cea1072caab6e8b21f3408575c  ...   \n",
       "\n",
       "                                         identifier_y       species     sex  \\\n",
       "0   [{'system': 'GDC', 'value': 'TCGA-EA-A6QX'}, {...  homo sapiens  female   \n",
       "1   [{'system': 'GDC', 'value': 'TCGA-EA-A6QX'}, {...  homo sapiens  female   \n",
       "2   [{'system': 'GDC', 'value': 'TCGA-EK-A2H1'}, {...  homo sapiens  female   \n",
       "3   [{'system': 'GDC', 'value': 'TCGA-EK-A2H1'}, {...  homo sapiens  female   \n",
       "4   [{'system': 'GDC', 'value': 'TCGA-IW-A3M4'}, {...  homo sapiens  female   \n",
       "5   [{'system': 'GDC', 'value': 'TCGA-AX-A2H5'}, {...  homo sapiens  female   \n",
       "6   [{'system': 'GDC', 'value': 'TCGA-AX-A2H5'}, {...  homo sapiens  female   \n",
       "7   [{'system': 'GDC', 'value': 'TCGA-FU-A3HZ'}, {...  homo sapiens  female   \n",
       "8   [{'system': 'GDC', 'value': 'TCGA-FU-A3HZ'}, {...  homo sapiens  female   \n",
       "9   [{'system': 'GDC', 'value': 'TCGA-VS-A9UL'}, {...  homo sapiens  female   \n",
       "10  [{'system': 'GDC', 'value': 'TCGA-VS-A9UL'}, {...  homo sapiens  female   \n",
       "11  [{'system': 'GDC', 'value': 'TCGA-VS-A9UP'}, {...  homo sapiens  female   \n",
       "12  [{'system': 'GDC', 'value': 'TCGA-VS-A9UP'}, {...  homo sapiens  female   \n",
       "13  [{'system': 'GDC', 'value': 'TCGA-B5-A0JN'}, {...  homo sapiens  female   \n",
       "14  [{'system': 'GDC', 'value': 'TCGA-B5-A0JN'}, {...  homo sapiens  female   \n",
       "\n",
       "                         race               ethnicity days_to_birth  \\\n",
       "0                       asian  not hispanic or latino      -18035.0   \n",
       "1                       asian  not hispanic or latino      -18035.0   \n",
       "2                       white            not reported       -7649.0   \n",
       "3                       white            not reported       -7649.0   \n",
       "4                       white  not hispanic or latino      -17721.0   \n",
       "5   black or african american  not hispanic or latino      -24757.0   \n",
       "6   black or african american  not hispanic or latino      -24757.0   \n",
       "7                       white  not hispanic or latino      -23565.0   \n",
       "8                       white  not hispanic or latino      -23565.0   \n",
       "9                       white            not reported      -29033.0   \n",
       "10                      white            not reported      -29033.0   \n",
       "11               not reported            not reported      -15849.0   \n",
       "12               not reported            not reported      -15849.0   \n",
       "13  black or african american            not reported      -30933.0   \n",
       "14  black or african american            not reported      -30933.0   \n",
       "\n",
       "   subject_associated_project vital_status days_to_death cause_of_death  \n",
       "0      [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "1      [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "2      [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "3      [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "4      [tcga_sarc, TCGA-SARC]        Alive           NaN           None  \n",
       "5      [TCGA-UCEC, tcga_ucec]        Alive           NaN           None  \n",
       "6      [TCGA-UCEC, tcga_ucec]        Alive           NaN           None  \n",
       "7      [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "8      [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "9      [TCGA-CESC, tcga_cesc]         Dead         442.0           None  \n",
       "10     [TCGA-CESC, tcga_cesc]         Dead         442.0           None  \n",
       "11     [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "12     [TCGA-CESC, tcga_cesc]        Alive           NaN           None  \n",
       "13     [TCGA-UCEC, tcga_ucec]        Alive           NaN           None  \n",
       "14     [TCGA-UCEC, tcga_ucec]        Alive           NaN           None  \n",
       "\n",
       "[15 rows x 32 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "slidemetadata = pd.merge(slidesdf, \n",
    "                         allmetadata, \n",
    "                         on=\"subject_id\")\n",
    "slidemetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3abb0f32",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "She saves this file out as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ef6e0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "slidemetadata.to_csv(\"slidemetadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "369a3068",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Now Julia has all the information she needs to begin work on her project. She can use the `drs_id` column information to directly download the images she is interested in using a DRS resolver, or she can input the DRS IDs at a cloud workspace such as [Terra](https://terra.bio/) or the [Cancer Genomics Cloud](https://www.cancergenomicscloud.org/) to view the images online. In either case, she has all the metadata she needs to get started, and can save this notebook of her work in case she'd like to come back and modify her search."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "4578c8680ee810f847df558484335f5ffb0f004d38a87276387030f59580c508"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
