{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "614a478f",
   "metadata": {},
   "source": [
    "# Cohort Building"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8004e87",
   "metadata": {},
   "source": [
    "**Example use case:** \n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" alt=\"alt_text\" align=\"left\"\n",
    "\twidth=\"150\" height=\"150\" />\n",
    "Julia is an oncologist that specializes in female reproductive health. As part of her research, she is interested in  using existing data on uterine cancers. If possible, she would like to see multiple datatypes (gross imaging, genomic data, proteomic data, histology) that come from the same patient, so she can look for shared phenotypes to test for their potential as early diagnostics. Julia heard that the Cancer Data Aggregator has made it easy to search across multiple datasets created by NCI, and so has decided to start her search there.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14e45a9b",
   "metadata": {},
   "source": [
    "## Getting Started\n",
    "\n",
    "The CDA provides a custom python tool for searching CDA data. [`Q`](usage/#q) (short for Query) offers several ways to search and filter data, and several input modes:\n",
    "\n",
    "---\n",
    "- **<a href=\"../../QuickStart/usage/#q\">Q.()</a>** builds a query that can be used by `run()` or `count()`\n",
    "- **<a href=\"../../QuickStart/usage/#qrun\">Q.run()</a>** returns data for the specified search \n",
    "- **<a href=\"../../QuickStart/usage/#qcount\">Q.count()</a>** returns summary information (counts) data that fit the specified search\n",
    "- **<a href=\"../../QuickStart/usage/#columns\">columns()</a>** returns entity field names\n",
    "- **<a href=\"../../QuickStart/usage/#unique_terms\">unique_terms()</a>** returns entity field contents\n",
    "\n",
    "---\n",
    "\n",
    "Before Julia does any work, she needs to import these functions cdapython.\n",
    "She'll also need to import [pandas](https://pandas.pydata.org/) to get nice dataframes.\n",
    "Finally, she tells cdapython to report it's version so she can be sure she's using the one she means to:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a5265d4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">2022.9</span>.<span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">15</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1;36m2022.9\u001b[0m.\u001b[1;36m15\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from cdapython import Q, columns, unique_terms, query\n",
    "import cdapython\n",
    "import pandas as pd \n",
    "print(cdapython.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c577d416",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'https://cancerdata.dsde-dev.broadinstitute.org/'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q.set_default_project_dataset(\"broad-dsde-dev.cda_dev\")\n",
    "Q.set_host_url(\"https://cancerdata.dsde-dev.broadinstitute.org/\")\n",
    "Q.get_host_url()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7a4e1b8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'broad-dsde-dev.cda_dev'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Q.get_default_project_dataset()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75eef23e",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "    \n",
    "CDA data comes from three sources:\n",
    "<ul>\n",
    "<li><b>The <a href=\"https://proteomic.datacommons.cancer.gov/pdc/\"> Proteomic Data Commons</a> (PDC)</b></li>\n",
    "<li><b>The <a href=\"https://gdc.cancer.gov/\">Genomic Data Commons</a> (GDC)</b></li>\n",
    "<li><b>The <a href=\"https://datacommons.cancer.gov/repository/imaging-data-commons\">Imaging Data Commons</a> (IDC)</b></li>\n",
    "</ul> \n",
    "    \n",
    "The CDA makes this data searchable in four main endpoints:\n",
    "\n",
    "<ul>\n",
    "<li><b>subject:</b> A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</li>\n",
    "<li><b>researchsubject:</b> A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</li>\n",
    "<li><b>specimen:</b> Any material taken as a sample from a biological entity (living or dead), or from a physical object or the environment. Specimens are usually collected as an example of their kind, often for use in some investigation.</li>\n",
    "<li><b>file:</b> A unit of data about subjects, researchsubjects, specimens, or their associated information</li>\n",
    "</ul>\n",
    "and two endpoints that offer deeper information about data in the researchsubject endpoint:\n",
    "<ul>\n",
    "<li><b>diagnosis:</b> A collection of characteristics that describe an abnormal condition of the body as assessed at a point in time. May be used to capture information about neoplastic and non-neoplastic conditions.</li>\n",
    "<li><b>treatment:</b> Represent medication administration or other treatment types.</li>\n",
    "</ul>\n",
    "Any metadata field can be searched from any endpoint, the only difference between search types is what type of data is returned by default. This means that you can think of the CDA as a really, really enormous spreadsheet full of data. To search this enormous spreadsheet, you'd want select columns, and then filter rows.\n",
    "</div>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "391bc9a7",
   "metadata": {},
   "source": [
    "## Finding Search Terms\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "   Accordingly, to see what search fields are available, Julia starts by using the command `columns`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ef0dd8e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\"><span style=\"font-weight: bold\">[</span>\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'AA_MAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in NHLBI-ESP African American population'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'AFR_MAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in 1000 Genomes combined African </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">population'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ALLELE_NUM'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Allele number from input; 0 is reference, 1 is first alternate etc.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'AMR_MAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in 1000 Genomes combined American </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">population'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Allele'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The variant allele used to calculate the consequence'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Amino_acids'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Amino acid substitution caused by the mutation. Only given if the variation affects the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">protein-coding sequence'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'BIOTYPE'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Biotype of transcript'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'CANONICAL'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'A flag (YES) indicating that the VEP-based canonical transcript, the longest translation, was</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">used for this gene. If not, the value is null'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'CCDS'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The  CCDS identifier for this transcript, where applicable'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'CDS_position'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Relative position of base pair in coding sequence. A - symbol is displayed as the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">numerator if the variant does not appear in coding sequence'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'CLIN_SIG'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Clinical significance of variant from dbSNP'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'CONTEXT'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The reference allele per VCF specs, and its five flanking base pairs'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'COSMIC'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Overlapping COSMIC variants'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Center'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'One or more genome sequencing center reporting the variant'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Chromosome'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Chromosome, possible values: chr1-22, and chrX'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Codons'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The alternative codons with the variant base in upper case'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Consequence'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Consequence type of this variant; sequence ontology terms'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'DISTANCE'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Shortest distance from the variant to transcript'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'DOMAINS'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The source and identifier of any overlapping protein domains'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'EAS_MAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in 1000 Genomes combined East Asian </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">population'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'EA_MAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in NHLBI-ESP European American </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">population'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ENSP'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The Ensembl protein identifier of the affected transcript'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'EUR_MAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in 1000 Genomes combined European </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">population'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'EXON'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The exon number (out of total number)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'End_Position'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Highest numeric genomic position of the reported variant on the genomic reference </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">sequence. Mutation end coordinate'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Entrez_Gene_Id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Entrez gene ID (an integer). 0 is used for regions that do not correspond to a gene </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">region or Ensembl ID'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Global Allele Frequency from   ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_AFR'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'African/African American Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_AMR'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'American Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_Adj'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Adjusted Global Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_EAS'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'East Asian Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_FIN'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Finnish Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_NFE'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-Finnish European Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_OTH'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Other Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'ExAC_AF_SAS'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'South Asian Allele Frequency from ExAC'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Existing_variation'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Known identifier of existing variation'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Exon_Number'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The exon number (out of total number)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'FILTER'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Copied from input VCF. This includes filters implemented directly by the variant caller and </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">other external software used in the DNA-Seq pipeline. See below for additional details.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Feature'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Stable Ensembl ID of feature (transcript, regulatory, motif)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Feature_type'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Type of feature. Currently one of Transcript, RegulatoryFeature, MotifFeature (or blank)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'GDC_FILTER'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'GDC filters applied universally across all MAFs'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'GDC_Validation_Status'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'GDC implementation of validation checks. See notes section (#5) below for </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">details'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'GMAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in   1000 Genomes'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Gene'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The gene symbol. In this table, gene symbol is gene name e.g. ACADVL'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'HGNC_ID'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Gene identifier from the HUGO Gene Nomenclature Committee if applicable'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'HGVS_OFFSET'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Indicates by how many bases the HGVS notations for this variant have been shifted'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'HGVSc'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The coding sequence of the variant in HGVS recommended format'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'HGVSp'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The protein sequence of the variant in HGVS recommended format. p.= signifies no change in the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">protein'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'HGVSp_Short'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Same as the HGVSp column, but using 1-letter amino-acid codes'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Hugo_Symbol'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'HUGO symbol for the gene (HUGO symbols are always in all caps). Unknown is used for regions</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">that do not correspond to a gene'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'IMPACT'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The impact modifier for the consequence type'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'INTRON'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The intron number (out of total number)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'MC3_Overlap'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Indicates whether this region overlaps with an MC3 variant for the same sample pair'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'MINIMISED'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Alleles in this variant have been converted to minimal representation before consequence </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">calculation (1 or null)'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Matched_Norm_Sample_UUID'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Unique GDC identifier for normal aliquot (10189 unique)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Mutation_Status'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'An assessment of the mutation as somatic, germline, LOH, post transcriptional </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">modification, unknown, or none. The values allowed in this field are constrained by the value in the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">Validation_Status field'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'NCBI_Build'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The reference genome used for the alignment (GRCh38)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'One_Consequence'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The single consequence of the canonical transcript in  sequence ontology terms, eg </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">missense_variant'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'PHENO'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Indicates if existing variant is associated with a phenotype, disease or trait (0, 1, or null)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'PICK'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"Indicates if this block of consequence data was picked by VEP's   pick feature (1 or null)\"</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'PUBMED'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Pubmed ID(s) of publications that cite existing variant'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'PolyPhen'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The PolyPhen prediction and/or score'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Protein_position'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Relative position of affected amino acid in protein. A - symbol is displayed as the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">numerator if the variant does not appear in coding sequence'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'RefSeq'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'RefSeq identifier for this transcript'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Reference_Allele'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The plus strand reference allele at this position. Includes the deleted sequence for a</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">deletion or - for an insertion'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'SAS_MAF'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Non-reference allele and frequency of existing variant in 1000 Genomes combined South Asian </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">population'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'SIFT'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The   SIFT prediction and/or score, with both given as prediction (score)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'SOMATIC'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Somatic status of each ID reported under Existing_variation (0, 1, or null)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'SWISSPROT'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'UniProtKB/Swiss-Prot accession'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'SYMBOL'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Eg TP53, LRP1B, etc (same as Hugo_Symbol field except blank instead of Unknown'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'SYMBOL_SOURCE'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The source of the gene symbol, usually HGNC, rarely blank, other sources include </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">Uniprot_gn, EntrezGene, etc'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Sequencer'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Instrument used to produce primary sequence data'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Start_Position'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Lowest numeric position of the reported variant on the genomic reference sequence. </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">Mutation start coordinate'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Strand'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Either + or - to denote whether read mapped to the sense (+) or anti-sense (-) strand'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'TRANSCRIPT_STRAND'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The DNA strand (1 or -1) on which the transcript/feature lies'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'TREMBL'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'UniProtKB/TrEMBL identifier of protein product'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'TSL'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Transcript support level, which is based on independent RNA analyses'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Transcript_ID'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Ensembl ID of the transcript affected by the variant'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Tumor_Sample_UUID'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Unique GDC identifier for tumor aliquot (10189 unique)'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Tumor_Seq_Allele1'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Primary data genotype for tumor sequencing (discovery) allele 1. A - symbol for a </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">deletion represents a variant. A - symbol for an insertion represents wild-type allele. Novel inserted sequence for</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">insertion does not include flanking reference bases'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Tumor_Seq_Allele2'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Primary data genotype for tumor sequencing (discovery) allele 2. A - symbol for a </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">deletion represents a variant. A - symbol for an insertion represents wild-type allele. Novel inserted sequence for</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">insertion does not include flanking reference bases'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Tumor_Validation_Allele1'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Secondary data from orthogonal technology. Tumor genotyping (validation) for </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">allele 1. A - symbol for a deletion represents a variant. A - symbol for an insertion represents wild-type allele. </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">Novel inserted sequence for insertion does not include flanking reference bases'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Tumor_Validation_Allele2'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Secondary data from orthogonal technology. Tumor genotyping (validation) for </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">allele 2'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'UNIPARC'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'UniParc identifier of protein product'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'VARIANT_CLASS'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Sequence Ontology variant class'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Validation_Method'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The assay platforms used for the validation call'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'Variant_Classification'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Translational effect of variant allele'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'Variant_Type'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Type of mutation. TNP (tri-nucleotide polymorphism) is analogous to DNP (di-nucleotide </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">polymorphism) but for three consecutive nucleotides. ONP (oligo-nucleotide polymorphism) is analogous to TNP but </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">for consecutive runs of four or more (SNP, DNP, TNP, ONP, INS, DEL, or Consolidated)'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'age_at_diagnosis'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The age in days of the individual at the time of diagnosis'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'aliquot_barcode_normal'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'TCGA aliquot barcode for the normal control, eg TCGA-12-1089-01A-01D-0517-01'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'aliquot_barcode_tumor'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'TCGA aliquot barcode for the tumor, eg TCGA-12-1089-01A-01D-0517-01'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'all_effects'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'A semicolon delimited list of all possible variant effects, sorted by priority </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">([Symbol,Consequence,HGVSp_Short,Transcript_ID,RefSeq,HGVSc,Impact,Canonical,Sift,PolyPhen,Strand])'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'anatomical_site'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Per GDC Dictionary, the text term that represents the name of the primary disease site </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">of the submitted tumor sample; recommend dropping tumor; biospecimen_anatomic_site.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'byte_size'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Size of the file in bytes. Maps to dcat:byteSize.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'cDNA_position'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Relative position of base pair in the cDNA sequence as a fraction. A - symbol is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">displayed as the numerator if the variant does not appear in cDNA'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'callerName'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'|-delimited list of mutation caller(s) that agreed on this particular call, always in </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">alphabetical order: muse, mutect, somaticsniper, varscan'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'case_barcode'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Original TCGA case barcode, eg TCGA-DX-A8BN'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'case_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Unique GDC identifier for the underlying case'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'cause_of_death'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Coded value indicating the circumstance or condition that results in the death of the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">subject.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'checksum'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'A digit representing the sum of the correct digits in a piece of stored or transmitted digital</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">data, against which later comparisons can be made to detect errors in the data.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'data_category'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Broad categorization of the contents of the data file.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'data_modality'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Data modality describes the biological nature of the information gathered as the result </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">of an Activity, independent of the technology or methods used to produce the information.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'data_type'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Specific content type of the data file.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'days_to_birth'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"Number of days between the date used for index and the date from a person's date of birth</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">represented as a calculated negative number of days.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'days_to_collection'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The number of days from the index date to either the date a sample was collected for</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">a specific study or project, or the date a patient underwent a procedure (e.g. surgical resection) yielding a </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">sample that was eventually used for research.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'days_to_death'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"Number of days between the date used for index and the date from a person's date of death</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">represented as a calculated number of days.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'days_to_treatment_end'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">' The timepoint at which the treatment ended.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'days_to_treatment_start'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The timepoint at which the treatment started.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'dbSNP_RS'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The rs-IDs from the   dbSNP database, novel if not found in any database used, or null if </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">there is no dbSNP record, but it is found in other databases'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'dbSNP_Val_Status'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The dbSNP validation status is reported as a semicolon-separated list of statuses. The</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">union of all rs-IDs is taken when there are multiple'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'dbgap_accession_number'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The dbgap accession number for the project.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'derived_from_specimen'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'A source/parent specimen from which this one was directly derived.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'derived_from_subject'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The Patient/ResearchSubject, or Biologically Derived Materal (e.g. a cell line, </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">tissue culture, organoid) from which the specimen was directly or indirectly derived.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'diagnosis_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The 'logical' identifier of the entity in the repository, e.g. a UUID.  This 'id' is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">unique within a given system. The identified entity may have a different 'id' in a different system.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'diagnosis_identifier_system'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The system or namespace that defines the identifier.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'diagnosis_identifier_value'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The value of the identifier, as defined by the system.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'drs_uri'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'A string of characters used to identify a resource on the Data Repo Service(DRS).'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'ethnicity'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"An individual's self-described social and cultural grouping, specifically whether an </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">the U.S. Office of Management and Business and used by the U.S. Census Bureau.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'fileName'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'|-delimited list of name of underlying MAF file'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'fileUUID'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'|-delimited list of unique GDC identifiers for underlying MAF file'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'file_associated_project'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'A reference to the Project(s) of which this ResearchSubject is a member. The </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">associated_project may be embedded using the $ref definition or may be a reference to the id for the Project - or a</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">URI expressed as a string to an existing entity.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'file_format'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Format of the data files.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'file_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The 'logical' identifier of the entity in the repository, e.g. a UUID.  This 'id' is unique </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">within a given system. The identified entity may have a different 'id' in a different system.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'file_identifier_system'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The system or namespace that defines the identifier.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'file_identifier_value'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The value of the identifier, as defined by the system.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'grade'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The degree of abnormality of cancer cells, a measure of differentiation, the extent to which </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">cancer cells are similar in appearance and function to healthy cells of the same tissue type. The degree of </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">differentiation often relates to the clinical behavior of the particular tumor. Based on the microscopic findings, </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">tumor grade is commonly described by one of four degrees of severity. Histopathologic grade of a tumor may be used </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">to plan treatment and estimate the future course, outcome, and overall prognosis of disease. Certain types of </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">cancers, such as soft tissue sarcoma, primary brain tumors, lymphomas, and breast have special grading systems.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'imaging_modality'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'An imaging modality describes the imaging equipment and/or method used to acquire </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">certain structural or functional information about the body. These include but are not limited to computed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'imaging_series'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The 'logical' identifier of the series or grouping of imaging files in the system of </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">record which the file is a part of.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'label'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Short name or abbreviation for dataset. Maps to rdfs:label.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'member_of_research_project'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'A reference to the Study(s) of which this ResearchSubject is a member.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'method_of_diagnosis'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The method used to confirm the patients malignant diagnosis'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'morphology'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Code that represents the histology of the disease using the third edition of the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">International Classification of Diseases for Oncology, published in 2000, used principally in tumor and cancer </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">registri'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'n_depth'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Read depth across this locus in normal BAM'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'normal_bam_uuid'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Unique GDC identifier for the underlying normal bam file'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'number_of_cycles'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The number of treatment cycles the subject received.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'primary_diagnosis'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The diagnosis instance that qualified a subject for inclusion on a ResearchProject'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'primary_diagnosis_condition'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The text term used to describe the type of malignant disease, as </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">categorized by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'primary_diagnosis_site'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The text term used to describe the primary site of disease, as categorized by </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O). This </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">categorization groups cases into general categories.  This attribute represents the primary site of disease that </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">qualified the subject for inclusion on the ResearchProject.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'primary_disease_type'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The text term used to describe the type of malignant disease, as categorized by </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).   This </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">attribute represents the disease that qualified the subject for inclusion on the ResearchProject.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'project_short_name'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Project name abbreviation; the program name appended with a project name </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">abbreviation; eg. TCGA-OV, etc.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'race'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'An arbitrary classification of a taxonomic group that is a division of a species. It usually </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">arises as a consequence of geographical isolation within a species and is characterized by shared heredity, </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business</span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">and used by the U.S. Census Bureau.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'researchsubject_Files'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'List of ids of File entities associated with the ResearchSubject'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'researchsubject_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The 'logical' identifier of the entity in the system of record, e.g. a UUID.  This </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">CDA, this is case_id.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'researchsubject_identifier_system'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The system or namespace that defines the identifier.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'researchsubject_identifier_value'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The value of the identifier, as defined by the system.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'sample_barcode_normal'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'TCGA sample barcode for the normal control, eg TCGA-12-1089-01A. One sample may </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">have multiple sets of CN segmentations corresponding to multiple aliquots; use GROUP BY appropriately in queries'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'sample_barcode_tumor'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'TCGA sample barcode for the tumor, eg TCGA-12-1089-01A. One sample may have </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">multiple sets of CN segmentations corresponding to multiple aliquots; use GROUP BY appropriately in queries'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'sex'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The biologic character or quality that distinguishes male and female from one another as expressed </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">characteristics.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'source_material_type'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The general kind of material from which the specimen was derived, indicating the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">physical nature of the source material. '</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'species'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The taxonomic group (e.g. species) of the patient. For MVP, since taxonomy vocabulary is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">consistent between GDC and PDC, using text.  Ultimately, this will be a term returned by the vocabulary service.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'specimen_Files'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'List of ids of File entities associated with the Specimen'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'specimen_associated_project'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The Project associated with the specimen.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'specimen_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The 'logical' identifier of the entity in the system of record, e.g. a UUID.  This 'id' is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">unique within a given system. The identified entity may have a different 'id' in a different system.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'specimen_identifier_system'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The system or namespace that defines the identifier.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'specimen_identifier_value'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The value of the identifier, as defined by the system.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'specimen_type'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The high-level type of the specimen, based on its how it has been derived from the </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">original extracted sample. \\n'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'src_vcf_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'|-delimited list of GDC VCF file identifiers'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'stage'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The extent of a cancer in the body. Staging is usually based on the size of the tumor, whether </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">lymph nodes contain cancer, and whether the cancer has spread from the original site to other parts of the body.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'subject_Files'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'List of ids of File entities associated with the Patient'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'subject_associated_project'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The list of Projects associated with the Subject.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'subject_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The 'logical' identifier of the entity in the system of record, e.g. a UUID.  This 'id' is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">unique within a given system. The identified entity may have a different 'id' in a different system.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'subject_identifier_system'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The system or namespace that defines the identifier.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'subject_identifier_value'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The value of the identifier, as defined by the system.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'t_alt_count'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Read depth supporting the variant allele in tumor BAM'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'t_depth'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Read depth across this locus in tumor BAM'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'t_ref_count'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Read depth supporting the reference allele in tumor BAM'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'therapeutic_agent'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'One or more therapeutic agents as part of this treatment.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'treatment_anatomic_site'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The anatomical site that the treatment targets.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'treatment_effect'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The effect of a treatment on the diagnosis or tumor.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'treatment_end_reason'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The reason the treatment ended.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'treatment_id'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">\"The 'logical' identifier of the entity in the repository, e.g. a UUID.  This 'id' is </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">unique within a given system. The identified entity may have a different 'id' in a different system.\"</span>\n",
       "    <span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'treatment_identifier_system'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The system or namespace that defines the identifier.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'treatment_identifier_value'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The value of the identifier, as defined by the system.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'treatment_outcome'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The final outcome of the treatment.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'treatment_type'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'The treatment type including medication/therapeutics or other procedures.'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span><span style=\"color: #008000; text-decoration-color: #008000\">'tumor_bam_uuid'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Unique GDC identifier for the underlying bam file'</span><span style=\"font-weight: bold\">}</span>,\n",
       "    <span style=\"font-weight: bold\">{</span>\n",
       "        <span style=\"color: #008000; text-decoration-color: #008000\">'vital_status'</span>: <span style=\"color: #008000; text-decoration-color: #008000\">'Coded value indicating the state or condition of being living or deceased; also includes </span>\n",
       "<span style=\"color: #008000; text-decoration-color: #008000\">the case where the vital status is unknown.'</span>\n",
       "    <span style=\"font-weight: bold\">}</span>\n",
       "<span style=\"font-weight: bold\">]</span>\n",
       "</pre>\n"
      ],
      "text/plain": [
       "\u001b[1m[\u001b[0m\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'AA_MAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in NHLBI-ESP African American population'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'AFR_MAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in 1000 Genomes combined African \u001b[0m\n",
       "\u001b[32mpopulation'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ALLELE_NUM'\u001b[0m: \u001b[32m'Allele number from input; 0 is reference, 1 is first alternate etc.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'AMR_MAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in 1000 Genomes combined American \u001b[0m\n",
       "\u001b[32mpopulation'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Allele'\u001b[0m: \u001b[32m'The variant allele used to calculate the consequence'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Amino_acids'\u001b[0m: \u001b[32m'Amino acid substitution caused by the mutation. Only given if the variation affects the \u001b[0m\n",
       "\u001b[32mprotein-coding sequence'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'BIOTYPE'\u001b[0m: \u001b[32m'Biotype of transcript'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'CANONICAL'\u001b[0m: \u001b[32m'A flag \u001b[0m\u001b[32m(\u001b[0m\u001b[32mYES\u001b[0m\u001b[32m)\u001b[0m\u001b[32m indicating that the VEP-based canonical transcript, the longest translation, was\u001b[0m\n",
       "\u001b[32mused for this gene. If not, the value is null'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'CCDS'\u001b[0m: \u001b[32m'The  CCDS identifier for this transcript, where applicable'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'CDS_position'\u001b[0m: \u001b[32m'Relative position of base pair in coding sequence. A - symbol is displayed as the \u001b[0m\n",
       "\u001b[32mnumerator if the variant does not appear in coding sequence'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'CLIN_SIG'\u001b[0m: \u001b[32m'Clinical significance of variant from dbSNP'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'CONTEXT'\u001b[0m: \u001b[32m'The reference allele per VCF specs, and its five flanking base pairs'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'COSMIC'\u001b[0m: \u001b[32m'Overlapping COSMIC variants'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Center'\u001b[0m: \u001b[32m'One or more genome sequencing center reporting the variant'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Chromosome'\u001b[0m: \u001b[32m'Chromosome, possible values: chr1-22, and chrX'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Codons'\u001b[0m: \u001b[32m'The alternative codons with the variant base in upper case'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Consequence'\u001b[0m: \u001b[32m'Consequence type of this variant; sequence ontology terms'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'DISTANCE'\u001b[0m: \u001b[32m'Shortest distance from the variant to transcript'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'DOMAINS'\u001b[0m: \u001b[32m'The source and identifier of any overlapping protein domains'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'EAS_MAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in 1000 Genomes combined East Asian \u001b[0m\n",
       "\u001b[32mpopulation'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'EA_MAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in NHLBI-ESP European American \u001b[0m\n",
       "\u001b[32mpopulation'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ENSP'\u001b[0m: \u001b[32m'The Ensembl protein identifier of the affected transcript'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'EUR_MAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in 1000 Genomes combined European \u001b[0m\n",
       "\u001b[32mpopulation'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'EXON'\u001b[0m: \u001b[32m'The exon number \u001b[0m\u001b[32m(\u001b[0m\u001b[32mout of total number\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'End_Position'\u001b[0m: \u001b[32m'Highest numeric genomic position of the reported variant on the genomic reference \u001b[0m\n",
       "\u001b[32msequence. Mutation end coordinate'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Entrez_Gene_Id'\u001b[0m: \u001b[32m'Entrez gene ID \u001b[0m\u001b[32m(\u001b[0m\u001b[32man integer\u001b[0m\u001b[32m)\u001b[0m\u001b[32m. 0 is used for regions that do not correspond to a gene \u001b[0m\n",
       "\u001b[32mregion or Ensembl ID'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF'\u001b[0m: \u001b[32m'Global Allele Frequency from   ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_AFR'\u001b[0m: \u001b[32m'African/African American Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_AMR'\u001b[0m: \u001b[32m'American Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_Adj'\u001b[0m: \u001b[32m'Adjusted Global Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_EAS'\u001b[0m: \u001b[32m'East Asian Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_FIN'\u001b[0m: \u001b[32m'Finnish Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_NFE'\u001b[0m: \u001b[32m'Non-Finnish European Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_OTH'\u001b[0m: \u001b[32m'Other Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'ExAC_AF_SAS'\u001b[0m: \u001b[32m'South Asian Allele Frequency from ExAC'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Existing_variation'\u001b[0m: \u001b[32m'Known identifier of existing variation'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Exon_Number'\u001b[0m: \u001b[32m'The exon number \u001b[0m\u001b[32m(\u001b[0m\u001b[32mout of total number\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'FILTER'\u001b[0m: \u001b[32m'Copied from input VCF. This includes filters implemented directly by the variant caller and \u001b[0m\n",
       "\u001b[32mother external software used in the DNA-Seq pipeline. See below for additional details.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Feature'\u001b[0m: \u001b[32m'Stable Ensembl ID of feature \u001b[0m\u001b[32m(\u001b[0m\u001b[32mtranscript, regulatory, motif\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Feature_type'\u001b[0m: \u001b[32m'Type of feature. Currently one of Transcript, RegulatoryFeature, MotifFeature \u001b[0m\u001b[32m(\u001b[0m\u001b[32mor blank\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'GDC_FILTER'\u001b[0m: \u001b[32m'GDC filters applied universally across all MAFs'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'GDC_Validation_Status'\u001b[0m: \u001b[32m'GDC implementation of validation checks. See notes section \u001b[0m\u001b[32m(\u001b[0m\u001b[32m#5\u001b[0m\u001b[32m)\u001b[0m\u001b[32m below for \u001b[0m\n",
       "\u001b[32mdetails'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'GMAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in   1000 Genomes'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Gene'\u001b[0m: \u001b[32m'The gene symbol. In this table, gene symbol is gene name e.g. ACADVL'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'HGNC_ID'\u001b[0m: \u001b[32m'Gene identifier from the HUGO Gene Nomenclature Committee if applicable'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'HGVS_OFFSET'\u001b[0m: \u001b[32m'Indicates by how many bases the HGVS notations for this variant have been shifted'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'HGVSc'\u001b[0m: \u001b[32m'The coding sequence of the variant in HGVS recommended format'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'HGVSp'\u001b[0m: \u001b[32m'The protein sequence of the variant in HGVS recommended format. p.= signifies no change in the \u001b[0m\n",
       "\u001b[32mprotein'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'HGVSp_Short'\u001b[0m: \u001b[32m'Same as the HGVSp column, but using 1-letter amino-acid codes'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Hugo_Symbol'\u001b[0m: \u001b[32m'HUGO symbol for the gene \u001b[0m\u001b[32m(\u001b[0m\u001b[32mHUGO symbols are always in all caps\u001b[0m\u001b[32m)\u001b[0m\u001b[32m. Unknown is used for regions\u001b[0m\n",
       "\u001b[32mthat do not correspond to a gene'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'IMPACT'\u001b[0m: \u001b[32m'The impact modifier for the consequence type'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'INTRON'\u001b[0m: \u001b[32m'The intron number \u001b[0m\u001b[32m(\u001b[0m\u001b[32mout of total number\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'MC3_Overlap'\u001b[0m: \u001b[32m'Indicates whether this region overlaps with an MC3 variant for the same sample pair'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'MINIMISED'\u001b[0m: \u001b[32m'Alleles in this variant have been converted to minimal representation before consequence \u001b[0m\n",
       "\u001b[32mcalculation \u001b[0m\u001b[32m(\u001b[0m\u001b[32m1 or null\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Matched_Norm_Sample_UUID'\u001b[0m: \u001b[32m'Unique GDC identifier for normal aliquot \u001b[0m\u001b[32m(\u001b[0m\u001b[32m10189 unique\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Mutation_Status'\u001b[0m: \u001b[32m'An assessment of the mutation as somatic, germline, LOH, post transcriptional \u001b[0m\n",
       "\u001b[32mmodification, unknown, or none. The values allowed in this field are constrained by the value in the \u001b[0m\n",
       "\u001b[32mValidation_Status field'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'NCBI_Build'\u001b[0m: \u001b[32m'The reference genome used for the alignment \u001b[0m\u001b[32m(\u001b[0m\u001b[32mGRCh38\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'One_Consequence'\u001b[0m: \u001b[32m'The single consequence of the canonical transcript in  sequence ontology terms, eg \u001b[0m\n",
       "\u001b[32mmissense_variant'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'PHENO'\u001b[0m: \u001b[32m'Indicates if existing variant is associated with a phenotype, disease or trait \u001b[0m\u001b[32m(\u001b[0m\u001b[32m0, 1, or null\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'PICK'\u001b[0m: \u001b[32m\"Indicates if this block of consequence data was picked by VEP's   pick feature \u001b[0m\u001b[32m(\u001b[0m\u001b[32m1 or null\u001b[0m\u001b[32m)\u001b[0m\u001b[32m\"\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'PUBMED'\u001b[0m: \u001b[32m'Pubmed ID\u001b[0m\u001b[32m(\u001b[0m\u001b[32ms\u001b[0m\u001b[32m)\u001b[0m\u001b[32m of publications that cite existing variant'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'PolyPhen'\u001b[0m: \u001b[32m'The PolyPhen prediction and/or score'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Protein_position'\u001b[0m: \u001b[32m'Relative position of affected amino acid in protein. A - symbol is displayed as the \u001b[0m\n",
       "\u001b[32mnumerator if the variant does not appear in coding sequence'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'RefSeq'\u001b[0m: \u001b[32m'RefSeq identifier for this transcript'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Reference_Allele'\u001b[0m: \u001b[32m'The plus strand reference allele at this position. Includes the deleted sequence for a\u001b[0m\n",
       "\u001b[32mdeletion or - for an insertion'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'SAS_MAF'\u001b[0m: \u001b[32m'Non-reference allele and frequency of existing variant in 1000 Genomes combined South Asian \u001b[0m\n",
       "\u001b[32mpopulation'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'SIFT'\u001b[0m: \u001b[32m'The   SIFT prediction and/or score, with both given as prediction \u001b[0m\u001b[32m(\u001b[0m\u001b[32mscore\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'SOMATIC'\u001b[0m: \u001b[32m'Somatic status of each ID reported under Existing_variation \u001b[0m\u001b[32m(\u001b[0m\u001b[32m0, 1, or null\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'SWISSPROT'\u001b[0m: \u001b[32m'UniProtKB/Swiss-Prot accession'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'SYMBOL'\u001b[0m: \u001b[32m'Eg TP53, LRP1B, etc \u001b[0m\u001b[32m(\u001b[0m\u001b[32msame as Hugo_Symbol field except blank instead of Unknown'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'SYMBOL_SOURCE'\u001b[0m: \u001b[32m'The source of the gene symbol, usually HGNC, rarely blank, other sources include \u001b[0m\n",
       "\u001b[32mUniprot_gn, EntrezGene, etc'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Sequencer'\u001b[0m: \u001b[32m'Instrument used to produce primary sequence data'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Start_Position'\u001b[0m: \u001b[32m'Lowest numeric position of the reported variant on the genomic reference sequence. \u001b[0m\n",
       "\u001b[32mMutation start coordinate'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Strand'\u001b[0m: \u001b[32m'Either + or - to denote whether read mapped to the sense \u001b[0m\u001b[32m(\u001b[0m\u001b[32m+\u001b[0m\u001b[32m)\u001b[0m\u001b[32m or anti-sense \u001b[0m\u001b[32m(\u001b[0m\u001b[32m-\u001b[0m\u001b[32m)\u001b[0m\u001b[32m strand'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'TRANSCRIPT_STRAND'\u001b[0m: \u001b[32m'The DNA strand \u001b[0m\u001b[32m(\u001b[0m\u001b[32m1 or -1\u001b[0m\u001b[32m)\u001b[0m\u001b[32m on which the transcript/feature lies'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'TREMBL'\u001b[0m: \u001b[32m'UniProtKB/TrEMBL identifier of protein product'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'TSL'\u001b[0m: \u001b[32m'Transcript support level, which is based on independent RNA analyses'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Transcript_ID'\u001b[0m: \u001b[32m'Ensembl ID of the transcript affected by the variant'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Tumor_Sample_UUID'\u001b[0m: \u001b[32m'Unique GDC identifier for tumor aliquot \u001b[0m\u001b[32m(\u001b[0m\u001b[32m10189 unique\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Tumor_Seq_Allele1'\u001b[0m: \u001b[32m'Primary data genotype for tumor sequencing \u001b[0m\u001b[32m(\u001b[0m\u001b[32mdiscovery\u001b[0m\u001b[32m)\u001b[0m\u001b[32m allele 1. A - symbol for a \u001b[0m\n",
       "\u001b[32mdeletion represents a variant. A - symbol for an insertion represents wild-type allele. Novel inserted sequence for\u001b[0m\n",
       "\u001b[32minsertion does not include flanking reference bases'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Tumor_Seq_Allele2'\u001b[0m: \u001b[32m'Primary data genotype for tumor sequencing \u001b[0m\u001b[32m(\u001b[0m\u001b[32mdiscovery\u001b[0m\u001b[32m)\u001b[0m\u001b[32m allele 2. A - symbol for a \u001b[0m\n",
       "\u001b[32mdeletion represents a variant. A - symbol for an insertion represents wild-type allele. Novel inserted sequence for\u001b[0m\n",
       "\u001b[32minsertion does not include flanking reference bases'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Tumor_Validation_Allele1'\u001b[0m: \u001b[32m'Secondary data from orthogonal technology. Tumor genotyping \u001b[0m\u001b[32m(\u001b[0m\u001b[32mvalidation\u001b[0m\u001b[32m)\u001b[0m\u001b[32m for \u001b[0m\n",
       "\u001b[32mallele 1. A - symbol for a deletion represents a variant. A - symbol for an insertion represents wild-type allele. \u001b[0m\n",
       "\u001b[32mNovel inserted sequence for insertion does not include flanking reference bases'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Tumor_Validation_Allele2'\u001b[0m: \u001b[32m'Secondary data from orthogonal technology. Tumor genotyping \u001b[0m\u001b[32m(\u001b[0m\u001b[32mvalidation\u001b[0m\u001b[32m)\u001b[0m\u001b[32m for \u001b[0m\n",
       "\u001b[32mallele 2'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'UNIPARC'\u001b[0m: \u001b[32m'UniParc identifier of protein product'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'VARIANT_CLASS'\u001b[0m: \u001b[32m'Sequence Ontology variant class'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Validation_Method'\u001b[0m: \u001b[32m'The assay platforms used for the validation call'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'Variant_Classification'\u001b[0m: \u001b[32m'Translational effect of variant allele'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'Variant_Type'\u001b[0m: \u001b[32m'Type of mutation. TNP \u001b[0m\u001b[32m(\u001b[0m\u001b[32mtri-nucleotide polymorphism\u001b[0m\u001b[32m)\u001b[0m\u001b[32m is analogous to DNP \u001b[0m\u001b[32m(\u001b[0m\u001b[32mdi-nucleotide \u001b[0m\n",
       "\u001b[32mpolymorphism\u001b[0m\u001b[32m)\u001b[0m\u001b[32m but for three consecutive nucleotides. ONP \u001b[0m\u001b[32m(\u001b[0m\u001b[32moligo-nucleotide polymorphism\u001b[0m\u001b[32m)\u001b[0m\u001b[32m is analogous to TNP but \u001b[0m\n",
       "\u001b[32mfor consecutive runs of four or more \u001b[0m\u001b[32m(\u001b[0m\u001b[32mSNP, DNP, TNP, ONP, INS, DEL, or Consolidated\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'age_at_diagnosis'\u001b[0m: \u001b[32m'The age in days of the individual at the time of diagnosis'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'aliquot_barcode_normal'\u001b[0m: \u001b[32m'TCGA aliquot barcode for the normal control, eg TCGA-12-1089-01A-01D-0517-01'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'aliquot_barcode_tumor'\u001b[0m: \u001b[32m'TCGA aliquot barcode for the tumor, eg TCGA-12-1089-01A-01D-0517-01'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'all_effects'\u001b[0m: \u001b[32m'A semicolon delimited list of all possible variant effects, sorted by priority \u001b[0m\n",
       "\u001b[32m(\u001b[0m\u001b[32m[\u001b[0m\u001b[32mSymbol,Consequence,HGVSp_Short,Transcript_ID,RefSeq,HGVSc,Impact,Canonical,Sift,PolyPhen,Strand\u001b[0m\u001b[32m]\u001b[0m\u001b[32m)\u001b[0m\u001b[32m'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'anatomical_site'\u001b[0m: \u001b[32m'Per GDC Dictionary, the text term that represents the name of the primary disease site \u001b[0m\n",
       "\u001b[32mof the submitted tumor sample; recommend dropping tumor; biospecimen_anatomic_site.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'byte_size'\u001b[0m: \u001b[32m'Size of the file in bytes. Maps to dcat:byteSize.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'cDNA_position'\u001b[0m: \u001b[32m'Relative position of base pair in the cDNA sequence as a fraction. A - symbol is \u001b[0m\n",
       "\u001b[32mdisplayed as the numerator if the variant does not appear in cDNA'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'callerName'\u001b[0m: \u001b[32m'|-delimited list of mutation caller\u001b[0m\u001b[32m(\u001b[0m\u001b[32ms\u001b[0m\u001b[32m)\u001b[0m\u001b[32m that agreed on this particular call, always in \u001b[0m\n",
       "\u001b[32malphabetical order: muse, mutect, somaticsniper, varscan'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'case_barcode'\u001b[0m: \u001b[32m'Original TCGA case barcode, eg TCGA-DX-A8BN'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'case_id'\u001b[0m: \u001b[32m'Unique GDC identifier for the underlying case'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'cause_of_death'\u001b[0m: \u001b[32m'Coded value indicating the circumstance or condition that results in the death of the \u001b[0m\n",
       "\u001b[32msubject.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'checksum'\u001b[0m: \u001b[32m'A digit representing the sum of the correct digits in a piece of stored or transmitted digital\u001b[0m\n",
       "\u001b[32mdata, against which later comparisons can be made to detect errors in the data.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'data_category'\u001b[0m: \u001b[32m'Broad categorization of the contents of the data file.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'data_modality'\u001b[0m: \u001b[32m'Data modality describes the biological nature of the information gathered as the result \u001b[0m\n",
       "\u001b[32mof an Activity, independent of the technology or methods used to produce the information.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'data_type'\u001b[0m: \u001b[32m'Specific content type of the data file.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'days_to_birth'\u001b[0m: \u001b[32m\"Number of days between the date used for index and the date from a person's date of birth\u001b[0m\n",
       "\u001b[32mrepresented as a calculated negative number of days.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'days_to_collection'\u001b[0m: \u001b[32m'The number of days from the index date to either the date a sample was collected for\u001b[0m\n",
       "\u001b[32ma specific study or project, or the date a patient underwent a procedure \u001b[0m\u001b[32m(\u001b[0m\u001b[32me.g. surgical resection\u001b[0m\u001b[32m)\u001b[0m\u001b[32m yielding a \u001b[0m\n",
       "\u001b[32msample that was eventually used for research.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'days_to_death'\u001b[0m: \u001b[32m\"Number of days between the date used for index and the date from a person's date of death\u001b[0m\n",
       "\u001b[32mrepresented as a calculated number of days.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'days_to_treatment_end'\u001b[0m: \u001b[32m' The timepoint at which the treatment ended.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'days_to_treatment_start'\u001b[0m: \u001b[32m'The timepoint at which the treatment started.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'dbSNP_RS'\u001b[0m: \u001b[32m'The rs-IDs from the   dbSNP database, novel if not found in any database used, or null if \u001b[0m\n",
       "\u001b[32mthere is no dbSNP record, but it is found in other databases'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'dbSNP_Val_Status'\u001b[0m: \u001b[32m'The dbSNP validation status is reported as a semicolon-separated list of statuses. The\u001b[0m\n",
       "\u001b[32munion of all rs-IDs is taken when there are multiple'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'dbgap_accession_number'\u001b[0m: \u001b[32m'The dbgap accession number for the project.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'derived_from_specimen'\u001b[0m: \u001b[32m'A source/parent specimen from which this one was directly derived.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'derived_from_subject'\u001b[0m: \u001b[32m'The Patient/ResearchSubject, or Biologically Derived Materal \u001b[0m\u001b[32m(\u001b[0m\u001b[32me.g. a cell line, \u001b[0m\n",
       "\u001b[32mtissue culture, organoid\u001b[0m\u001b[32m)\u001b[0m\u001b[32m from which the specimen was directly or indirectly derived.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'diagnosis_id'\u001b[0m: \u001b[32m\"The 'logical' identifier of the entity in the repository, e.g. a UUID.  This 'id' is \u001b[0m\n",
       "\u001b[32munique within a given system. The identified entity may have a different 'id' in a different system.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'diagnosis_identifier_system'\u001b[0m: \u001b[32m'The system or namespace that defines the identifier.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'diagnosis_identifier_value'\u001b[0m: \u001b[32m'The value of the identifier, as defined by the system.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'drs_uri'\u001b[0m: \u001b[32m'A string of characters used to identify a resource on the Data Repo Service\u001b[0m\u001b[32m(\u001b[0m\u001b[32mDRS\u001b[0m\u001b[32m)\u001b[0m\u001b[32m.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'ethnicity'\u001b[0m: \u001b[32m\"An individual's self-described social and cultural grouping, specifically whether an \u001b[0m\n",
       "\u001b[32mindividual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by \u001b[0m\n",
       "\u001b[32mthe U.S. Office of Management and Business and used by the U.S. Census Bureau.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'fileName'\u001b[0m: \u001b[32m'|-delimited list of name of underlying MAF file'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'fileUUID'\u001b[0m: \u001b[32m'|-delimited list of unique GDC identifiers for underlying MAF file'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'file_associated_project'\u001b[0m: \u001b[32m'A reference to the Project\u001b[0m\u001b[32m(\u001b[0m\u001b[32ms\u001b[0m\u001b[32m)\u001b[0m\u001b[32m of which this ResearchSubject is a member. The \u001b[0m\n",
       "\u001b[32massociated_project may be embedded using the $ref definition or may be a reference to the id for the Project - or a\u001b[0m\n",
       "\u001b[32mURI expressed as a string to an existing entity.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'file_format'\u001b[0m: \u001b[32m'Format of the data files.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'file_id'\u001b[0m: \u001b[32m\"The 'logical' identifier of the entity in the repository, e.g. a UUID.  This 'id' is unique \u001b[0m\n",
       "\u001b[32mwithin a given system. The identified entity may have a different 'id' in a different system.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'file_identifier_system'\u001b[0m: \u001b[32m'The system or namespace that defines the identifier.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'file_identifier_value'\u001b[0m: \u001b[32m'The value of the identifier, as defined by the system.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'grade'\u001b[0m: \u001b[32m'The degree of abnormality of cancer cells, a measure of differentiation, the extent to which \u001b[0m\n",
       "\u001b[32mcancer cells are similar in appearance and function to healthy cells of the same tissue type. The degree of \u001b[0m\n",
       "\u001b[32mdifferentiation often relates to the clinical behavior of the particular tumor. Based on the microscopic findings, \u001b[0m\n",
       "\u001b[32mtumor grade is commonly described by one of four degrees of severity. Histopathologic grade of a tumor may be used \u001b[0m\n",
       "\u001b[32mto plan treatment and estimate the future course, outcome, and overall prognosis of disease. Certain types of \u001b[0m\n",
       "\u001b[32mcancers, such as soft tissue sarcoma, primary brain tumors, lymphomas, and breast have special grading systems.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'imaging_modality'\u001b[0m: \u001b[32m'An imaging modality describes the imaging equipment and/or method used to acquire \u001b[0m\n",
       "\u001b[32mcertain structural or functional information about the body. These include but are not limited to computed \u001b[0m\n",
       "\u001b[32mtomography \u001b[0m\u001b[32m(\u001b[0m\u001b[32mCT\u001b[0m\u001b[32m)\u001b[0m\u001b[32m and magnetic resonance imaging \u001b[0m\u001b[32m(\u001b[0m\u001b[32mMRI\u001b[0m\u001b[32m)\u001b[0m\u001b[32m. Taken from the DICOM standard.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'imaging_series'\u001b[0m: \u001b[32m\"The 'logical' identifier of the series or grouping of imaging files in the system of \u001b[0m\n",
       "\u001b[32mrecord which the file is a part of.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'label'\u001b[0m: \u001b[32m'Short name or abbreviation for dataset. Maps to rdfs:label.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'member_of_research_project'\u001b[0m: \u001b[32m'A reference to the Study\u001b[0m\u001b[32m(\u001b[0m\u001b[32ms\u001b[0m\u001b[32m)\u001b[0m\u001b[32m of which this ResearchSubject is a member.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'method_of_diagnosis'\u001b[0m: \u001b[32m'The method used to confirm the patients malignant diagnosis'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'morphology'\u001b[0m: \u001b[32m'Code that represents the histology of the disease using the third edition of the \u001b[0m\n",
       "\u001b[32mInternational Classification of Diseases for Oncology, published in 2000, used principally in tumor and cancer \u001b[0m\n",
       "\u001b[32mregistri'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'n_depth'\u001b[0m: \u001b[32m'Read depth across this locus in normal BAM'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'normal_bam_uuid'\u001b[0m: \u001b[32m'Unique GDC identifier for the underlying normal bam file'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'number_of_cycles'\u001b[0m: \u001b[32m'The number of treatment cycles the subject received.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'primary_diagnosis'\u001b[0m: \u001b[32m'The diagnosis instance that qualified a subject for inclusion on a ResearchProject'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'primary_diagnosis_condition'\u001b[0m: \u001b[32m\"The text term used to describe the type of malignant disease, as \u001b[0m\n",
       "\u001b[32mcategorized by the World Health Organization's \u001b[0m\u001b[32m(\u001b[0m\u001b[32mWHO\u001b[0m\u001b[32m)\u001b[0m\u001b[32m International Classification of Diseases for Oncology \u001b[0m\u001b[32m(\u001b[0m\u001b[32mICD-O\u001b[0m\u001b[32m)\u001b[0m\u001b[32m.\u001b[0m\n",
       "\u001b[32mThis attribute represents the disease that qualified the subject for inclusion on the ResearchProject.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'primary_diagnosis_site'\u001b[0m: \u001b[32m\"The text term used to describe the primary site of disease, as categorized by \u001b[0m\n",
       "\u001b[32mthe World Health Organization's \u001b[0m\u001b[32m(\u001b[0m\u001b[32mWHO\u001b[0m\u001b[32m)\u001b[0m\u001b[32m International Classification of Diseases for Oncology \u001b[0m\u001b[32m(\u001b[0m\u001b[32mICD-O\u001b[0m\u001b[32m)\u001b[0m\u001b[32m. This \u001b[0m\n",
       "\u001b[32mcategorization groups cases into general categories.  This attribute represents the primary site of disease that \u001b[0m\n",
       "\u001b[32mqualified the subject for inclusion on the ResearchProject.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'primary_disease_type'\u001b[0m: \u001b[32m\"The text term used to describe the type of malignant disease, as categorized by \u001b[0m\n",
       "\u001b[32mthe World Health Organization's \u001b[0m\u001b[32m(\u001b[0m\u001b[32mWHO\u001b[0m\u001b[32m)\u001b[0m\u001b[32m International Classification of Diseases for Oncology \u001b[0m\u001b[32m(\u001b[0m\u001b[32mICD-O\u001b[0m\u001b[32m)\u001b[0m\u001b[32m.   This \u001b[0m\n",
       "\u001b[32mattribute represents the disease that qualified the subject for inclusion on the ResearchProject.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'project_short_name'\u001b[0m: \u001b[32m'Project name abbreviation; the program name appended with a project name \u001b[0m\n",
       "\u001b[32mabbreviation; eg. TCGA-OV, etc.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'race'\u001b[0m: \u001b[32m'An arbitrary classification of a taxonomic group that is a division of a species. It usually \u001b[0m\n",
       "\u001b[32marises as a consequence of geographical isolation within a species and is characterized by shared heredity, \u001b[0m\n",
       "\u001b[32mphysical attributes and behavior, and in the case of humans, by common history, nationality, or geographic \u001b[0m\n",
       "\u001b[32mdistribution. The provided values are based on the categories defined by the U.S. Office of Management and Business\u001b[0m\n",
       "\u001b[32mand used by the U.S. Census Bureau.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'researchsubject_Files'\u001b[0m: \u001b[32m'List of ids of File entities associated with the ResearchSubject'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'researchsubject_id'\u001b[0m: \u001b[32m\"The 'logical' identifier of the entity in the system of record, e.g. a UUID.  This \u001b[0m\n",
       "\u001b[32m'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For \u001b[0m\n",
       "\u001b[32mCDA, this is case_id.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'researchsubject_identifier_system'\u001b[0m: \u001b[32m'The system or namespace that defines the identifier.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'researchsubject_identifier_value'\u001b[0m: \u001b[32m'The value of the identifier, as defined by the system.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'sample_barcode_normal'\u001b[0m: \u001b[32m'TCGA sample barcode for the normal control, eg TCGA-12-1089-01A. One sample may \u001b[0m\n",
       "\u001b[32mhave multiple sets of CN segmentations corresponding to multiple aliquots; use GROUP BY appropriately in queries'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'sample_barcode_tumor'\u001b[0m: \u001b[32m'TCGA sample barcode for the tumor, eg TCGA-12-1089-01A. One sample may have \u001b[0m\n",
       "\u001b[32mmultiple sets of CN segmentations corresponding to multiple aliquots; use GROUP BY appropriately in queries'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'sex'\u001b[0m: \u001b[32m\"The biologic character or quality that distinguishes male and female from one another as expressed \u001b[0m\n",
       "\u001b[32mby analysis of the person's gonadal, morphologic \u001b[0m\u001b[32m(\u001b[0m\u001b[32minternal and external\u001b[0m\u001b[32m)\u001b[0m\u001b[32m, chromosomal, and hormonal \u001b[0m\n",
       "\u001b[32mcharacteristics.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'source_material_type'\u001b[0m: \u001b[32m'The general kind of material from which the specimen was derived, indicating the \u001b[0m\n",
       "\u001b[32mphysical nature of the source material. '\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'species'\u001b[0m: \u001b[32m'The taxonomic group \u001b[0m\u001b[32m(\u001b[0m\u001b[32me.g. species\u001b[0m\u001b[32m)\u001b[0m\u001b[32m of the patient. For MVP, since taxonomy vocabulary is \u001b[0m\n",
       "\u001b[32mconsistent between GDC and PDC, using text.  Ultimately, this will be a term returned by the vocabulary service.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'specimen_Files'\u001b[0m: \u001b[32m'List of ids of File entities associated with the Specimen'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'specimen_associated_project'\u001b[0m: \u001b[32m'The Project associated with the specimen.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'specimen_id'\u001b[0m: \u001b[32m\"The 'logical' identifier of the entity in the system of record, e.g. a UUID.  This 'id' is \u001b[0m\n",
       "\u001b[32munique within a given system. The identified entity may have a different 'id' in a different system.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'specimen_identifier_system'\u001b[0m: \u001b[32m'The system or namespace that defines the identifier.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'specimen_identifier_value'\u001b[0m: \u001b[32m'The value of the identifier, as defined by the system.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'specimen_type'\u001b[0m: \u001b[32m'The high-level type of the specimen, based on its how it has been derived from the \u001b[0m\n",
       "\u001b[32moriginal extracted sample. \\n'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'src_vcf_id'\u001b[0m: \u001b[32m'|-delimited list of GDC VCF file identifiers'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'stage'\u001b[0m: \u001b[32m'The extent of a cancer in the body. Staging is usually based on the size of the tumor, whether \u001b[0m\n",
       "\u001b[32mlymph nodes contain cancer, and whether the cancer has spread from the original site to other parts of the body.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'subject_Files'\u001b[0m: \u001b[32m'List of ids of File entities associated with the Patient'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'subject_associated_project'\u001b[0m: \u001b[32m'The list of Projects associated with the Subject.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'subject_id'\u001b[0m: \u001b[32m\"The 'logical' identifier of the entity in the system of record, e.g. a UUID.  This 'id' is \u001b[0m\n",
       "\u001b[32munique within a given system. The identified entity may have a different 'id' in a different system.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'subject_identifier_system'\u001b[0m: \u001b[32m'The system or namespace that defines the identifier.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'subject_identifier_value'\u001b[0m: \u001b[32m'The value of the identifier, as defined by the system.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m't_alt_count'\u001b[0m: \u001b[32m'Read depth supporting the variant allele in tumor BAM'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m't_depth'\u001b[0m: \u001b[32m'Read depth across this locus in tumor BAM'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m't_ref_count'\u001b[0m: \u001b[32m'Read depth supporting the reference allele in tumor BAM'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'therapeutic_agent'\u001b[0m: \u001b[32m'One or more therapeutic agents as part of this treatment.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'treatment_anatomic_site'\u001b[0m: \u001b[32m'The anatomical site that the treatment targets.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'treatment_effect'\u001b[0m: \u001b[32m'The effect of a treatment on the diagnosis or tumor.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'treatment_end_reason'\u001b[0m: \u001b[32m'The reason the treatment ended.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'treatment_id'\u001b[0m: \u001b[32m\"The 'logical' identifier of the entity in the repository, e.g. a UUID.  This 'id' is \u001b[0m\n",
       "\u001b[32munique within a given system. The identified entity may have a different 'id' in a different system.\"\u001b[0m\n",
       "    \u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'treatment_identifier_system'\u001b[0m: \u001b[32m'The system or namespace that defines the identifier.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'treatment_identifier_value'\u001b[0m: \u001b[32m'The value of the identifier, as defined by the system.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'treatment_outcome'\u001b[0m: \u001b[32m'The final outcome of the treatment.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'treatment_type'\u001b[0m: \u001b[32m'The treatment type including medication/therapeutics or other procedures.'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\u001b[32m'tumor_bam_uuid'\u001b[0m: \u001b[32m'Unique GDC identifier for the underlying bam file'\u001b[0m\u001b[1m}\u001b[0m,\n",
       "    \u001b[1m{\u001b[0m\n",
       "        \u001b[32m'vital_status'\u001b[0m: \u001b[32m'Coded value indicating the state or condition of being living or deceased; also includes \u001b[0m\n",
       "\u001b[32mthe case where the vital status is unknown.'\u001b[0m\n",
       "    \u001b[1m}\u001b[0m\n",
       "\u001b[1m]\u001b[0m\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(columns().to_list())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd05eba2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "   \n",
    "There are a lot of columns in the CDA data, but Julia is most interested in diagnosis data, so she filters the list to only those:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "536970c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'age_at_diagnosis': 'The age in days of the individual at the time of diagnosis'},\n",
       " {'diagnosis_id': \"The 'logical' identifier of the entity in the repository, e.g. a UUID.  This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.\"},\n",
       " {'diagnosis_identifier_system': 'The system or namespace that defines the identifier.'},\n",
       " {'diagnosis_identifier_value': 'The value of the identifier, as defined by the system.'},\n",
       " {'method_of_diagnosis': 'The method used to confirm the patients malignant diagnosis'},\n",
       " {'primary_diagnosis': 'The diagnosis instance that qualified a subject for inclusion on a ResearchProject'},\n",
       " {'primary_diagnosis_condition': \"The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).   This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.\"},\n",
       " {'primary_diagnosis_site': \"The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O). This categorization groups cases into general categories.  This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.\"}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns().to_list(filters=\"diagnosis\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63b4cf0",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "\n",
    "To search the CDA, a user also needs to know what search terms are available. Each column will contain a huge amount of data, so retreiving all of the rows would be overwhelming. Instead, the CDA has a `unique_terms()` function that will return all of the unique values that populate the requested column. Like `columns`, `unique_terms` defaults to giving us an overview of the results, and can be filtered.\n",
    "    \n",
    "</div>\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Since Julia is interested specificially in uterine cancers, she uses the `unique_terms` function to see what data is available for 'ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site' and 'ResearchSubject.primary_diagnosis_site' to see if 'uterine' appears:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4527dde5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre style=\"white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace\">\n",
       "            Http Status: <span style=\"color: #008080; text-decoration-color: #008080; font-weight: bold\">400</span>\n",
       "            Error Message: Column ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site does not exist\n",
       "            \n",
       "</pre>\n"
      ],
      "text/plain": [
       "\n",
       "            Http Status: \u001b[1;36m400\u001b[0m\n",
       "            Error Message: Column ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site does not exist\n",
       "            \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'to_list'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_74/2639024179.py\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0munique_terms\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'NoneType' object has no attribute 'to_list'"
     ]
    }
   ],
   "source": [
    "unique_terms(\"ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site\").to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "740e5955",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b005036b",
   "metadata": {},
   "source": [
    "<div class=\"cdanote\" style=\"background-color:#b3e5d5;color:black;padding:20px;\">\n",
    "    \n",
    "CDA makes multiple datasets searchable from a common interface, but does not harmonize the data. This means that researchers should review all the terms in a column, and not just choose the first one that fits, as there may be other similar terms available as well.\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73e6b8dc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia sees that \"treatment_anatomic_site\" does not have 'Uterine', but does have 'Cervix'. She also notes that both 'Uterus' and 'Uterus, NOS' are listed in the \"primary_diagnosis_site\" results. As she was initially looking for \"uterine\", Julia decides to expand her search a bit to account for variable naming schemes. So, she runs a fuzzy match filter on the \"ResearchSubject.primary_diagnosis_site\" for 'uter' as that should cover all variants:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31064125",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"uter\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9311a49e",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Just to be sure, Julia also searches for any other instances of \"cervix\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2038a8cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_terms(\"ResearchSubject.primary_diagnosis_site\").to_list(filters=\"cerv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29c4de58",
   "metadata": {},
   "source": [
    "## Building a Query\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "With all her likely terms found, Julia begins to create a search that will get data for all of her terms. She does this by writing a series of `Q` statements that define what rows should be returned from each column. For the \"treatment_anatomic_site\", only one term is of interest, so she uses the `=` operator to get only exact matches:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "951fcc8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Tsite = Q('ResearchSubject.Diagnosis.Treatment.treatment_anatomic_site = \"Cervix\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12cb5f72",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "However, for \"primary_diagnosis_site\", Julia has several terms she wants to search with. Luckily, `Q` also can run fuzzy searches. It can also search more than one term at a time, so Julia writes one big `Q` statement to grab everything that is either 'uter' or 'cerv':"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36cfd8a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "Dsite = Q('ResearchSubject.primary_diagnosis_site = \"%uter%\" OR ResearchSubject.primary_diagnosis_site = \"%cerv%\"')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "349af6f2",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia adds her two queries together into one large one:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f5f9e4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDATA = Tsite.OR(Dsite)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1f5cb55",
   "metadata": {},
   "source": [
    "## Looking at Summary Data\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Now that Julia has a query, she can use it to look for data in any of the CDA endpoints. She starts by getting an overall summary of what data is available using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "355b1706",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDATA.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7ce25fc",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "It seems there's a lot of data that might work for Julias study! Since she is interested in the beginings of cancer, she decides to start by looking at  the researchsubject information, since that is where most of the diagnosis information is. She again gets a summary using `count`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55b0cdeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ALLDATA.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a323e2",
   "metadata": {},
   "source": [
    "## Refining Queries\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Browsing the primary_diagnosis_condition data, Julia notices that there are a large number of research subjects that are Adenomas and Adenocarcinomas. Since Julia wants to look for common phenotypes in early cancers, she decides it might be easier to exclude the endocrine related data, as they might have different mechanisms. So she adds a new filter to her query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d526198",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "Noadeno = Q('ResearchSubject.primary_diagnosis_condition != \"Adenomas and Adenocarcinomas\"')\n",
    "\n",
    "NoAdenoData = ALLDATA.AND(Noadeno)\n",
    "\n",
    "NoAdenoData.researchsubject.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40a0191d",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "She then previews the actual metadata for researchsubject, subject, and file, to make sure that they have all the information she will need for her work. Since she's mostly interested in looking at the kinds of data available from each endpoint:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d186b837",
   "metadata": {},
   "outputs": [],
   "source": [
    "NoAdenoData.researchsubject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "086697b3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>ResearchSubject Field Definitions</h3>\n",
    "\n",
    "<i>A research subject is the entity of interest in a specific research study or project, typically a human being or an animal, but can also be a device, group of humans or animals, or a tissue sample. Human research subjects are usually not traceable to a particular person to protect the subjects privacy. This entity plays the role of the case_id in existing data. An individual who participates in 3 studies will have 3 researchsubject IDs</i>\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. For CDA, this is case_id.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>member_of_research_project:</b> A reference to the Study(s) of which this ResearchSubject is a member.</li>\n",
    "<li><b>primary_diagnosis_condition:</b> The text term used to describe the type of malignant disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This attribute represents the disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>primary_diagnosis_site:</b> The text term used to describe the primary site of disease, as categorized by the World Health Organization's (WHO) <a href=\"https://www.who.int/standards/classifications/other-classifications/international-classification-of-diseases-for-oncology\">International Classification of Diseases for Oncology</a> (ICD-O). This categorization groups cases into general categories. This attribute represents the primary site of disease that qualified the subject for inclusion on the ResearchProject.</li>\n",
    "<li><b>subject_id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system. Can be joined to the `id` field from subject results</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d0f5e2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "NoAdenoData.subject.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dec76132",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>Subject Field Definitions</h3>\n",
    "\n",
    "<i>A patient entity captures the study-independent metadata for research subjects. Human research subjects are usually not traceable to a particular person to protect the subjects privacy.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "<li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.\",STRING</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "<li><b>species:</b> The taxonomic group (e.g. species) of the patient. For MVP, since taxonomy vocabulary is consistent between GDC and PDC, using text. Ultimately, this will be a term returned by the vocabulary service.</li>\n",
    "<li><b>sex:</b> The biologic character or quality that distinguishes male and female from one another as expressed by analysis of the person's gonadal, morphologic (internal and external), chromosomal, and hormonal characteristics.</li>\n",
    "<li><b>race:</b> An arbitrary classification of a taxonomic group that is a division of a species. It usually arises as a consequence of geographical isolation within a species and is characterized by shared heredity, physical attributes and behavior, and in the case of humans, by common history, nationality, or geographic distribution. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>ethnicity:</b> An individual's self-described social and cultural grouping, specifically whether an individual describes themselves as Hispanic or Latino. The provided values are based on the categories defined by the U.S. Office of Management and Business and used by the U.S. Census Bureau.</li>\n",
    "<li><b>days_to_birth:</b> Number of days between the date used for index and the date from a person's date of birth represented as a calculated negative number of days.</li>\n",
    "<li><b>subject_associated_project:</b> The list of Projects associated with the Subject.</li>\n",
    "<li><b>vital_status:</b> Coded value indicating the state or condition of being living or deceased; also includes the case where the vital status is unknown.</li>\n",
    "<li><b>days_to_death:</b> Number of days between the date used for index and the date from a person's date of death represented as a calculated number of days.</li>\n",
    "<li><b>cause_of_death:</b> Coded value indicating the circumstance or condition that results in the death of the subject.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04e04136",
   "metadata": {},
   "outputs": [],
   "source": [
    "NoAdenoData.file.run().to_dataframe() # view the dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cf9f2d3",
   "metadata": {},
   "source": [
    "\n",
    "---\n",
    "\n",
    "<div class=\"cdadefine\" style=\"background-color:#add9e5;color:black;padding:20px;\">\n",
    "\n",
    "<h3>File Field Definitions</h3>\n",
    "\n",
    "<i>A file is an information-bearing electronic object that contains a physical embodiment of some information using a particular character encoding.</i>\n",
    "\n",
    "    \n",
    "<ul>\n",
    "  <li><b>id:</b> The 'logical' identifier of the entity in the system of record, e.g. a UUID. This 'id' is unique within a given system. The identified entity may have a different 'id' in a different system.</li>\n",
    "<li><b>identifier:</b> A 'business' identifier for the entity, typically as provided by an external system or authority, that persists across implementing systems (i.e. a 'logical' identifier). Uses a specialized, complex 'Identifier' data type to capture information about the source of the business identifier - or a URI expressed as a string to an existing entity.</li>\n",
    "<li><b>identifier.system:</b> The system or namespace that defines the identifier.</li>\n",
    "<li><b>identifier.value:</b> The value of the identifier, as defined by the system.</li>\n",
    "  <li><b>label:</b> Short name or abbreviation for dataset. Maps to rdfs:label.</li>\n",
    "  <li><b>data_catagory:</b> Broad categorization of the contents of the data file.</li>\n",
    "  <li><b>data_type:</b> Specific content type of the data file.</li>\n",
    "  <li><b>file_format:</b> Format of the data files.</li>\n",
    "  <li><b>associated_project:</b> A reference to the Project(s) of which this ResearchSubject is a member. The associated_project may be embedded using the ref definition or may be a reference to the id for the Project - or a URI expressed as a string to an existing entity.</li>\n",
    "  <li><b>drs_uri:</b> A string of characters used to identify a resource on the Data Repo Service(DRS). Can be used to retreive this specific file from a server.</li>\n",
    "  <li><b>byte_size:</b> Size of the file in bytes. Maps to dcat:byteSize.</li>\n",
    "  <li><b>checksum:</b> The md5 value for the file. A digit representing the sum of the correct digits in a piece of stored or transmitted digital data, against which later comparisons can be made to detect errors in the data.</li>\n",
    "  <li><b>data_modality:</b> Data modality describes the biological nature of the information gathered as the result of an Activity, independent of the technology or methods used to produce the information. Always one of \"Genomic\", \"Proteomic\", or \"Imaging\".</li>\n",
    "  <li><b>imaging_modality:</b> An imaging modality describes the imaging equipment and/or method used to acquire certain structural or functional information about the body. These include but are not limited to computed tomography (CT) and magnetic resonance imaging (MRI). Taken from the DICOM standard.</li>\n",
    "  <li><b>dbgap_accession_number:</b> The dbgap accession number for the project.</li>\n",
    "</ul>  \n",
    "\n",
    "</div>\n",
    "    \n",
    "---\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6aadbe",
   "metadata": {},
   "source": [
    "## Working with Results (pagination)\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Finally, Julia wants to save these results to use for the future. Since the preview dataframes only show the first 100 results of each search, she uses the `paginator` function to get all the data from the subject and researchsubject endpoints into their own dataframes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2cec2bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "researchsubs = NoAdenoData.researchsubject.run()\n",
    "rsdf = pd.DataFrame()\n",
    "for i in researchsubs.paginator(to_df=True):\n",
    "    rsdf = pd.concat([rsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1258057",
   "metadata": {},
   "outputs": [],
   "source": [
    "subs = NoAdenoData.subject.run()\n",
    "subsdf = pd.DataFrame()\n",
    "for i in subs.paginator(to_df=True):\n",
    "    subsdf = pd.concat([subsdf, i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04cd73df",
   "metadata": {},
   "outputs": [],
   "source": [
    "rsdf # view the researchsubject dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92a6f811",
   "metadata": {},
   "outputs": [],
   "source": [
    "subsdf # view the subject dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75bcbe86",
   "metadata": {},
   "source": [
    "## Merging Results across Endpoints\n",
    "\n",
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Then Julia uses the `id` fields in each result to merge them together into one big dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b7a3383",
   "metadata": {},
   "outputs": [],
   "source": [
    "allmetadata = pd.merge(rsdf,\n",
    "                subsdf,\n",
    "                left_on=\"subject_id\",\n",
    "                right_on='id')\n",
    "\n",
    "allmetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "024da831",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "And saves it out to a csv so she can browse it with Excel:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6628de4",
   "metadata": {},
   "outputs": [],
   "source": [
    "allmetadata.to_csv(\"allmetadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246644d3",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "   \n",
    "Julia knows from her subject count summary that there are more than 200,000 files associated with her subjects, which is likely far more than she needs. To help her decide what files she wants, Julia uses endpoint chaining to get summary information about the files that are assigned to researchsubjects for her search criteria:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae1ae079",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "NoAdenoData.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "111f47c0",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Julia decides that a good place to start would be with Slide Images. There's only 1111, so she should be able to quickly scan through them over the next few days and see if they will be useful. So she adds one more filter on her search:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4a170b3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "JustSlides = Q('file.data_type = \"Slide Image\"')\n",
    "NoadenoJustSlides = NoAdenoData.AND(JustSlides)\n",
    "NoadenoJustSlides.researchsubject.file.count.run()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e385faf",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Finally, Julia uses the pagenation function again to get all the slide files, and merges her metadata file with this file information. This way she will be able to review what phenotypes each slide is associated with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f45f6503",
   "metadata": {},
   "outputs": [],
   "source": [
    "slides = NoadenoJustSlides.researchsubject.file.run()\n",
    "slidesdf = pd.DataFrame()\n",
    "for i in slides.paginator(to_df=True):\n",
    "    slidesdf = pd.concat([slidesdf, i])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d1f7a21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "slidemetadata = pd.merge(slidesdf, \n",
    "                         allmetadata, \n",
    "                         on=\"subject_id\")\n",
    "slidemetadata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3abb0f32",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "She saves this file out as well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ef6e0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "slidemetadata.to_csv(\"slidemetadata.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "369a3068",
   "metadata": {},
   "source": [
    "<img src=\"https://github.com/CancerDataAggregator/readthedocs/blob/main/docs/Examples/images/Julia.png?raw=true\" align=\"left\"\n",
    "\twidth=\"50\" height=\"50\" />\n",
    "    \n",
    "Now Julia has all the information she needs to begin work on her project. She can use the `drs_id` column information to directly download the images she is interested in using a DRS resolver, or she can input the DRS IDs at a cloud workspace such as [Terra](https://terra.bio/) or the [Cancer Genomics Cloud](https://www.cancergenomicscloud.org/) to view the images online. In either case, she has all the metadata she needs to get started, and can save this notebook of her work in case she'd like to come back and modify her search."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "4578c8680ee810f847df558484335f5ffb0f004d38a87276387030f59580c508"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
